---
title: 'Residency and exhaustion signatures for CD8 T cells in Zhang et al Smart-seq2 single cell data'
author: 
  - name: 'Momeneh (Sepideh) Foroutan'
    affiliation: 'Huntington Cancer Immunotherapy Lab'
    url: https://www.monash.edu/discovery-institute/huntington-lab
date: '09-06-2020'
output:
  rmdformats::readthedown:
    fig_width: 12
    fig_height: 6
    gallery: TRUE
    highlight: tango
    lightbox: TRUE
    self_contained: TRUE
    thumbnails: FALSE
    number_sections: TRUE	
    toc_depth: 4
    use_bookdown: TRUE
    code_folding: hide
  html_document2:
    df_print: paged
params:
  update_date: !r paste("Last updated on:", Sys.Date() )
editor_options: 
  chunk_output_type: inline
---
`r params$update_date`

<style>
body {
text-align: justify}
</style>


```{r knitr_init, echo=FALSE, results="asis", cache=FALSE, echo=F,  message=F, warning=F}
library(knitr)
library(rmdformats)

options(max.print = "75")

opts_chunk$set(
  # echo = FALSE,
  cache = TRUE,
  # prompt = FALSE,
  # tidy = FALSE,
  comment = NA,
  message = FALSE,
  warning = FALSE,
  # results = FALSE,
  root.dir = normalizePath("."))

opts_knit$set(width = 65)

```

# Set up and overview

This document contains the codes used to generate some of the results in the [Foroutan et al](https://www.biorxiv.org/content/10.1101/2020.12.19.423379v1) paper. We used the Smart-seq2 data from Zhang et al study in this document. The processed, filtered and batch corrected data were downloaded from GEO under accession number GSE146771 on the 9th of June 2020.

We start with setting up the paths, loading the required packages/libraries, and reading the data and signatures. Then we subset the data to only CD8 T cells, and obtain initial Res and Exh signatures in this subset. We then use the scores from initial signatures to define Res and Exh cell subsets and perform DE analysis and percentile comparisons to define the final CD8 Res and Exh signatures. Finally, we perform trajectory analysis based on the results of the final signatures. This document reproduces a part of Figure 1 (first column) in the manuscript, and uses some of the signatures listed in Table 1. 

```{r file set up, echo=F, results=F, message=F, warning=F}

library(DT)
library(Seurat)
library(SingleCellExperiment)
library(singscore)
library(slingshot)
library(tidyverse)
library(RColorBrewer)

mainDir <- getwd()
outPath <- "../output/CD8/"
figPath <- "../figure/CD8/"
dataPath <- "../data/"
scriptPath <- "../script/"

tpmPath <- "../data/GSE146771_Zhang_CRC/"
hpaGoUniprotPath <- "../data/HPA_GO_Uniprot/"
ccleLcmPath <- "../data/CCLE_LCM/"

# ifelse(!dir.exists(file.path(mainDir, outPath)), dir.create(file.path(mainDir, outPath)), FALSE)
# ifelse(!dir.exists(file.path(mainDir, figPath)), dir.create(file.path(mainDir, figPath)), FALSE)
# ifelse(!dir.exists(file.path(mainDir, dataPath)), dir.create(file.path(mainDir, dataPath)), FALSE)
# ifelse(!dir.exists(file.path(mainDir, scriptPath)), dir.create(file.path(mainDir, scriptPath)), FALSE)

options(digits = 3)

equal_breaks <- function(n = nBreak, s = scalingFactor, ...){
  function(x){
    # rescaling
    d <- s * diff(range(x)) / (1+2*s)
    round( seq(min(x)+d, max(x)-d, length=n), 2)
  }
}

nBreak = 3
scalingFactor = 0.05

textSize <- 1.2

currentTheme <- theme_minimal() +
  theme(
    # panel.background = element_blank()
    axis.title = element_text(size = rel(textSize)),
    axis.text = element_text(angle = 0, size = rel(textSize)),
    # strip.background = element_rect(colour = "#f0f0f0", fill = "#f0f0f0"),
    strip.text = element_text(size = rel(textSize)),
    axis.line = element_line(colour = "black", size = 0.5),
    legend.position = 'top',
    # legend.title = element_text(size = rel(textSize), face = "italic"),
    # legend.text = element_text(size = rel(textSize)),
    legend.key.size = unit(1, 'lines'),
    ## increase the line space
    plot.title = element_text(
      face = "bold",
      size = rel(textSize),
      hjust = 0.5
    )
  )

cols <-  c(
  brewer.pal(8, "Dark2")[-5],
  brewer.pal(10, "Paired"),
  brewer.pal(12, "Set3"),
  brewer.pal(9, "Blues")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Oranges")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Greens")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Purples")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Reds")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Greys")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "BuGn")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "PuRd")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "BuPu")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "YlGn")[c(8, 3, 7, 4, 6, 9, 5)])


source(paste0(scriptPath, "visReduction_function.R"))
```



# Read SMART-seq2 CD8 data
We read the SMART-seq2 data, and generate a Seurat object for it. The data is already in TPM format.
Note that among 18 patients in Zhang et al study, 3 of them are MSI-H (P0123, P0413 and P0825), and the other 15 patients are MSS. So we add these info to the annotation file and calculate the percent of mitochondrial genes.

```{r, cache= T}
## dim:15179 10469

crcSmart <-
   data.table::fread(
    paste0(tpmPath, "GSE146771_CRC.Leukocyte.Smart-seq2.TPM.txt")
  )

rnames <- crcSmart$V1
crcSmart <- crcSmart[, 2:ncol(crcSmart)]
rownames(crcSmart) <- rnames

crcSmart <- as.matrix(crcSmart)
rownames(crcSmart) <- rnames


metaSmart <-
  read.table(
    paste0(tpmPath, "GSE146771_CRC.Leukocyte.Smart-seq2.Metadata.txt"),
    header = T,
    sep = "\t"
  )

metaSmart$Sub_Cluster <- as.character(metaSmart$Sub_Cluster)
metaSmart$Sub_Cluster <- sapply(metaSmart$Sub_Cluster, function(x){
  unlist(strsplit(x, "_"))[2]
})


##----- assign colours to cells
immCells <- metaSmart$Sub_Cluster[grepl("CD8", metaSmart$Sub_Cluster)|
                                    grepl("NK", metaSmart$Sub_Cluster) |
                                    grepl("CD4", metaSmart$Sub_Cluster)|
                                    grepl("ILC3", metaSmart$Sub_Cluster)]

cellCols <- cols[1:length(unique(immCells))]
names(cellCols) <- unique(immCells)[order(unique(immCells))]
cd8Cols <- cellCols[grepl("CD8", 
                          names(cellCols))]

##----- geenrate Seurat object

subSmartCD8 <- crcSmart[, grepl('CD8', metaSmart$Sub_Cluster) ]
subMetaSmartCD8 <- metaSmart[grepl('CD8', metaSmart$Sub_Cluster), ]


# all(subMetaSmartCD8$CellName == colnames(subSmartCD8))
rownames(subMetaSmartCD8) <- subMetaSmartCD8$CellName

subMetaSmartCD8$Sub_Cluster <- as.character(subMetaSmartCD8$Sub_Cluster)

sdRow <- matrixStats::rowSds(subSmartCD8)
subSmartCD8 <- subSmartCD8[ sdRow!=0 , ]

cd8 <-
  CreateSeuratObject(
    counts = subSmartCD8,
    project = 'Zhang_SmartSeq_CD8',
    meta.data = subMetaSmartCD8,
    min.cells = 5
  )

cd8[['percent.mt']] <- PercentageFeatureSet(cd8, pattern = '^MT-')

cd8$MSI <- "MSS"
cd8$MSI[cd8$Sample %in% c("P0123", "P0413", "P0825")] <- "MSI-H"

## save the RDS file of the CD8
# saveRDS(cd8, paste0(dataPath, "GSE146771_Zhang_CRC/CD8_Zhang_Smartseq_Seurat.RDS"))
```

## Read GPCRs and TFs
In order to annotate genes based on whether or not they are TFs, GPCR or associated with GPCR signalling, we use the human protein atlas and get the GPCR and TF annotations from there. For GPCR, we also downloaded the GPCR related genes from QuickGO and selected the genes annotated as gpcr in uniprot too. 
```{r}
hpa <- read.delim(paste0(hpaGoUniprotPath, "proteinatlas.tsv"), header = T)

GPCRs <-
  hpa[grepl("coupled", hpa$Gene.description, ignore.case = T) |
        grepl("coupled", hpa$Protein.class, ignore.case = T), c(
          "Gene",
          "Ensembl",
          "Protein.class",
          "Gene.description",
          "Uniprot" ,
          "Blood.RNA...NK.cell..NX."
        )]

GPCRs <- GPCRs[order(GPCRs$Blood.RNA...NK.cell..NX., decreasing = T), ]


TFs <- hpa[grepl("Transcription factor", hpa$Gene.description, ignore.case = T) |
        grepl("Transcription factor", hpa$Protein.class, ignore.case = T), c(
          "Gene",
          "Ensembl",
          "Protein.class",
          "Gene.description",
          "Uniprot" ,
          "Blood.RNA...NK.cell..NX."
        )]
  
TFs <- TFs[order(TFs$Blood.RNA...NK.cell..NX., decreasing = T), ]
 
  
go <- read.delim(paste0(hpaGoUniprotPath, "QuickGO-annotations-1592213989958-20200615.tsv"), header = T)

go <- unique(go$SYMBOL)

## 780 genes common between these
# length(intersect(GPCRs$Gene, go))

## 2035 unique genes
gpr <- unique(c(as.character(GPCRs$Gene), as.character(go)))
gpr <- as.character(gpr)
gpr <- gpr[! grepl("Homo sapiens", gpr)]
gpr <- gpr[! grepl("human", gpr)]
## MGRF

uniprot <- read.csv(paste0(hpaGoUniprotPath, "7tmrlist_uniprot.csv"), 
                    stringsAsFactors = F)

uniprot <- uniprot[grepl("HUMAN", uniprot$genes), ]
uniprot <- sapply(uniprot$genes, function(x){ unlist(strsplit(x, "_"))[1]})

uniprot <- as.character(uniprot)


tars <- read.csv(paste0(hpaGoUniprotPath, "targets_and_families.csv"), 
                 stringsAsFactors = F)

tars <- tars$HGNC.symbol[tars$Type == "gpcr"]
tars <- tars[! tars == ""]
tars <- tars[!duplicated(tars)]

gpr <- unique(c(gpr, uniprot, tars))

```

# Data analysis and visualisations

## Initial Res and Exh signatures

### Define genes for CD8 analysis
Here we have put together all the signature from CD8 Trm/exhaustion/dysfunction that we collected from the literature (see Table 1 in the paper). These are 1502 genes, of which 1353 genes exist in the single cell data. These include some well known canonnical Res and Exh markers.
```{r}
TcellSigsList <-
  readRDS(paste0(
    dataPath,
    "Signatures_literature/Tcell_signature_list_literature.RDS"
  ))

CD8Genes <-
  unique(
    unlist(TcellSigsList)
  )

```

### Cross correlation between genes
We perform cross correlation between all the 1353 genes, and then focus on the correlation between the canonical markers of CD8 Res/Exh and other genes.
```{r}

subSmartCD8 <- as.matrix(GetAssayData(cd8))

## 1353 genes
CD8Expr <- subSmartCD8[rownames(subSmartCD8) %in% CD8Genes, ]

corMat <- cor(t(CD8Expr), method = "spearman")


## function for flattening the cor matrix
flattenCorrMatrix <- function(cormat
) {
  ut <- upper.tri(cormat)
  data.frame(
    row = rownames(cormat)[row(cormat)[ut]],
    column = rownames(cormat)[col(cormat)[ut]],
    cor  =(cormat)[ut]
  )
}

corStat <- flattenCorrMatrix(corMat)

```

Canonical Res and Exh markers for CD8:

* Res: "CD69", "ITGAE", "ITGA1", "RGS1"
* Exh: "HAVCR2", "PDCD1", "CTLA4", "LAYN", "CXCL13"

Regarding the Res genes, we select genes that have correlation more than 0.35 with canonical Res markers. Note that both ZNF683 and CXCR6 did not show correlations more than 0.35 with any of the ~1300 genes, so we did not include them as the canonical Res markers in our list. this results in 41 genes.

Regarding the Exh genes, we select genes that have correlation more than 0.25 with canonical Exh markers. Note that LAG3 did not show correlations more than 0.35 with any of the ~1300 genes, so we did not include that as the canonical Exh markers in our list. Also, CTLA4 had correlation more than 0.25 with only 2 genes (CXCL13 and CD82). this results in 33 genes.

As some genes are common between both Res and Exh genes obtained above, i.e. have correlations with both Res and Exh canonical markers, we remove them from both gene lists. These genes are: "CXCR6" "DUSP4" "PHLDA1" "RGS1" "SAMSN1"  "KIR2DL4" "ATP8B4".

```{r}
trmMinGenes <- c( "CD69", "ITGAE", "ITGA1", "RGS1")
exhMinGenes <- c("HAVCR2", "PDCD1", "CTLA4", "LAYN", "CXCL13")
```


```{r}
trmCors <- corStat %>% filter(cor > 0.35 & (row %in% trmMinGenes | column %in% trmMinGenes))
trmGenes <- unique(c(as.character(trmCors$row),
                     as.character(trmCors$column)))

texCors <- corStat %>% filter(cor > 0.25 & (row %in% exhMinGenes | column %in% exhMinGenes))
texGenes <- unique(c(as.character(texCors$row),
                     as.character(texCors$column)))

```

Initial Res genes, before removing genes common with Exh genes:
```{r}
trmGenes
```

Initial Exh genes, before removing genes common with Res genes:
```{r}
texGenes
```

Genes common between the Res and Exh gene lists:
```{r}
comGs <- intersect(texGenes, trmGenes)
comGs

trmGenes <- trmGenes[! trmGenes %in% comGs]
texGenes <- texGenes[! texGenes %in% comGs]
```

### Sparse canonical correlation
Here, we also perform sparse canonical correlation (CCA) to remove genes that cause correlations between Res and Exh gene sets, although these genes are not common between the two signatures. We visualise the cross-correlation between the Res and Exh genes before and after removing genes based on sparse CCA.

Based on the results of the CCA, we remove "ITGA1" "ITGAE" "RGS2" "LDLRAD4" "CAPG" from the initial Res gene list and "LAYN" "GOLIM4" "ENTPD1" from the initial Exh gene list.

```{r, fig.height=10, fig.width = 12, warning = F, message = F, cache = T,  fig.cap = "Cross-correlation between all initial Res and Exh signature genes before and after CCA and removal of genes causing high correlation between the two signatures"}
CD8Expr <- subSmartCD8[unique(c(trmGenes, texGenes)), ]

corMat <- cor(t(CD8Expr), method = "spearman")

markerCols <- c("gray90", "gray20")
names(markerCols) <- c("Res", "Exh")

sampleAnnHm <-
  ComplexHeatmap::HeatmapAnnotation(Marker = as.character(c(rep(
    "Res", length(trmGenes)
  ), rep(
    "Exh", length(texGenes)
  ))),
  col = list(Marker = markerCols))



ComplexHeatmap::Heatmap(corMat,
                        show_column_names = T, 
                        top_annotation = sampleAnnHm, 
                        col = circlize::colorRamp2(
                          c(min(corMat), 0, max(corMat)), c("yellow3", "white", "navy")), 
                        # col = viridis::plasma(100), 
                        name = "Rho")


library(PMA)

cca_cd8_init <- CCA(
  t(subSmartCD8[rownames(subSmartCD8) %in% trmGenes, ]), 
   t(subSmartCD8[rownames(subSmartCD8) %in% texGenes, ]))

trmCorGenes <- colnames(t(subSmartCD8[rownames(subSmartCD8) %in% trmGenes, ]))[which(cca_cd8_init$u[, 1] <= -0.1)]
# "ITGA1"   "ITGAE"   "RGS2"    "LDLRAD4" "CAPG"  

texCorGenes <- colnames(t(subSmartCD8[rownames(subSmartCD8) %in% texGenes, ]))[which(cca_cd8_init$v[, 1] <= -0.1)]
# "LAYN"   "GOLIM4" "ENTPD1"

##---- remove genes that have non-zero values in CCA:
trmGenes <- trmGenes[! trmGenes %in% trmCorGenes]
texGenes <- texGenes[! texGenes %in% texCorGenes]


CD8Expr <- subSmartCD8[unique(c(trmGenes,texGenes)), ]

corMat <- cor(t(CD8Expr), method = "spearman")


sampleAnnHm <-
  ComplexHeatmap::HeatmapAnnotation(Marker = as.character(c(rep(
    "Res", length(trmGenes)
  ), rep(
    "Exh", length(texGenes)
  ))),
  col = list(Marker = markerCols))


ComplexHeatmap::Heatmap(corMat,
                        show_column_names = T, 
                        top_annotation = sampleAnnHm, 
                        col = circlize::colorRamp2(
                          c(min(corMat), 0, max(corMat)), c("yellow3", "white", "navy")), 
                        # col = viridis::plasma(100), 
                        name = "Rho")

```


```{r}
saveRDS(list(trmGenes = trmGenes, texGenes = texGenes), 
        paste0(outPath, "Initial_Res_Exh_Genes_CD8_Corr_20200928.RDS"))

# trmExhGenes <- readRDS(paste0(outPath, "Initial_Trm_Exh_Genes_CD8_Corr_20200928.RDS"))
# 
# trmGenes <- trmExhGenes$trmGenes
# texGenes <- trmExhGenes$texGenes
```

### Standard Seurat pipeline 
Find highly variable genes, Scale data, Clustering and dim reduction
```{r, warning = F, message = F, cache = T}
cd8 <- FindVariableFeatures(cd8, selection.method = "vst", nfeatures = 2000)

all.genes <- rownames(cd8)
cd8 <- ScaleData(cd8, features = all.genes)

cd8 <- RunPCA(cd8, features = VariableFeatures(object = cd8), verbose = F)

# DimHeatmap(cd8, dims = 1:30, cells = 500, balanced = TRUE)
# ElbowPlot(cd8, ndims = 40)

cd8 <- FindNeighbors(cd8, dims = 1:15)
cd8 <- FindClusters(cd8, resolution = 0.3, random.seed = 20200811)

cd8 <- RunUMAP(cd8, dims = 1:15, n.neighbors = 30, min.dist = 0.3, seed.use = 20200811)
```

### Counts vs percent mt or number of features
```{r, fig.height = 4, fig.width = 9, fig.cap = "Relationship between percent mito genes or number of detected genes and library size (from Seurat)"}

plot1 <-
  FeatureScatter(cd8,
                 feature1 = "nCount_RNA",
                 feature2 = "percent.mt",
                 group.by = "Sub_Cluster", 
                 cols = brewer.pal(8, "Set2"))
plot2 <-
  FeatureScatter(cd8,
                 feature1 = "nCount_RNA",
                 feature2 = "nFeature_RNA",
                 group.by = "Sample", 
                 cols = brewer.pal(9, "Set1"))
plot1 + plot2

```


```{r, fig.height=8, fig.width = 4, fig.cap = "Number of features, library size and percent mito genes in CD8 cells across patients"}
VlnPlot(
  cd8,
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
  ncol = 1,
  group.by = "Sample",
  pt.size = 0, cols = brewer.pal(9, "Set1")
)
```

### Highly Variable genes
```{r, fig.cap = "Highly variable genes"}
# cd8 <- FindVariableFeatures(cd8, selection.method = "vst", nfeatures = 2000)

top10 <- head(VariableFeatures(cd8), 20)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(cd8)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```


```{r, fig.height = 6, fig.width = 6, fig.cap = "Gene loadings for PC1 and PC2"}
VizDimLoadings(cd8, dims = 1:2, reduction = "pca")
```


## Score SMART-seq2 CD8 T cells based on initial Res and Exh signatures
```{r, message = F, warning = F}
library(singscore)
rd <- rankGenes(as.matrix(GetAssayData(cd8, slot = "data", assay = "RNA")))
     
cd8@meta.data$CuratedTrmGenes <- simpleScore(rankData = rd, 
                              upSet = trmGenes, 
                              knownDirection = T, 
                              centerScore = F, 
                              dispersionFun = var)$TotalScore    
     
cd8@meta.data$CuratedExhGenes <- simpleScore(rankData = rd, 
                              upSet = texGenes, 
                              knownDirection = T, 
                              centerScore = F, 
                              dispersionFun = var)$TotalScore    
## Save data after scores
# saveRDS(cd8, file = paste0(outPath, "Zhang_SmartSeq_SeuratObj_CD8_20200928_UMAP_scores.RDS"))

```


### Scatterplot of scores (related to Figure 1)
We generate scatterplots of the Initial Res vs Exh scores. In addition to reproducing a part of Figure 1, we colour cells based on different available annotations.
```{r, fig.height = 5, fig.width = 5, fig.cap = "Scatter plot of scores comparing initial Residency and initial Exhaustion signature scores"}

cell_attrsCD8 <- cd8[[]]

lowTrmPct <- quantile(cell_attrsCD8$CuratedTrmGenes, probs = 0.6)
highTrmPct <- quantile(cell_attrsCD8$CuratedTrmGenes, probs = 0.85)

lowExhPct <- quantile(cell_attrsCD8$CuratedExhGenes, probs = 0.6)
highExhPct <- quantile(cell_attrsCD8$CuratedExhGenes, probs = 0.85)


cell_attrsCD8 %>%
  ggplot(.,
         aes(x = CuratedTrmGenes, y = CuratedExhGenes)) +
  geom_point(alpha = 0.8, size = 1, color = "gray50") +
  geom_hline(yintercept = lowExhPct,
             col = "gray10",
             lty = 5,
             size = 1) +
  geom_hline(yintercept = highExhPct,
             col = "gray10",
             size = 1) +
  geom_vline(xintercept = lowTrmPct ,
             col = "gray10",
             size = 1) +
  geom_vline(xintercept = highTrmPct,
             col = "gray10",
              lty = 5,
             size = 1) +
  xlab("Initial residency score") +
  ylab("Initial exhaustion score") +
  currentTheme +
  guides(color = guide_legend(nrow = 3,
                              override.aes = list(size = 3)))


cell_attrsCD8 %>%
  ggplot(.,
         aes(x = CuratedTrmGenes, y = CuratedExhGenes, color = Sub_Cluster)) +
  geom_point(alpha = 0.8, size = 1) +
  scale_color_manual(values = cd8Cols) +
  geom_hline(yintercept = lowExhPct,
             col = "gray10",
             lty = 5,
             size = 1) +
  geom_hline(yintercept = highExhPct,
             col = "gray10",
             size = 1) +
  geom_vline(xintercept = lowTrmPct ,
             col = "gray10",
             size = 1) +
  geom_vline(xintercept = highTrmPct,
             col = "gray10",
              lty = 5,
             size = 1) +
  xlab("Initial residency score") +
  ylab("Initial exhaustion score") +
  currentTheme +
  guides(color = guide_legend(nrow = 3,
                              override.aes = list(size = 3)))


cell_attrsCD8 %>%
  ggplot(.,
         aes(x = CuratedTrmGenes, y = CuratedExhGenes, color = Tissue)) +
  geom_point(alpha = 0.8, size = 1) +
  scale_color_manual(values = cols) +
  geom_hline(yintercept = lowExhPct,
             col = "gray10",
             lty = 5,
             size = 1) +
  geom_hline(yintercept = highExhPct,
             col = "gray10",
             size = 1) +
  geom_vline(xintercept = lowTrmPct ,
             col = "gray10",
             size = 1) +
  geom_vline(xintercept = highTrmPct,
             col = "gray10",
              lty = 5,
             size = 1) +
  xlab("Initial residency score") +
  ylab("Initial exhaustion score") +
  currentTheme +
  guides(color = guide_legend(nrow = 3,
                              override.aes = list(size = 3)))
 

cell_attrsCD8 %>%
  ggplot(.,
         aes(x = CuratedTrmGenes, y = CuratedExhGenes, color = MSI)) +
  geom_point(alpha = 0.8, size = 1) +
  scale_color_manual(values = cols) +
  geom_hline(yintercept = lowExhPct,
             col = "gray10",
             lty = 5,
             size = 1) +
  geom_hline(yintercept = highExhPct,
             col = "gray10",
             size = 1) +
  geom_vline(xintercept = lowTrmPct ,
             col = "gray10",
             size = 1) +
  geom_vline(xintercept = highTrmPct,
             col = "gray10",
              lty = 5,
             size = 1) +
  xlab("Initial residency score") +
  ylab("Initial exhaustion score") +
  currentTheme +
  guides(color = guide_legend(nrow = 3,
                              override.aes = list(size = 3)))
 

```


## Final CD8 Res and Exh Markers
Define cells that are more Res vs those that are more Exh based on the above visualisations and thresholds. We then perform DE analysis between the Res and Exh cells, followed by percentile comparisons for gene expression across the cell types.

We first remove cell types that are specific to blood to focus only on the tissue infiltrated cells.
```{r}
cell_attrsCD8_Tissue <-
  cell_attrsCD8 %>%
  ## remove all the peripheral blood cells
  filter(Tissue != "P") %>%
  ## remove a few cells that are mostly specific to blood
  filter(! Sub_Cluster %in% c(
    # "CD4-ANXA1",
    # "CD4-CCR7",
    # "CD4-FOXP3",
    # "CD4-GNLY",
    "CD8-GPR183",
    "CD8-LEF1"
  ))

##----------- In both Tum and normal
cell_attrsCD8_Tissue$TissueExhCells[cell_attrsCD8_Tissue$CuratedExhGenes >
                                      highExhPct &
                                      cell_attrsCD8_Tissue$CuratedTrmGenes <
                                      lowTrmPct] <-
  "CD8_Exh"


cell_attrsCD8_Tissue$TissueExhCells[cell_attrsCD8_Tissue$CuratedTrmGenes >
                                      highTrmPct &
                                      cell_attrsCD8_Tissue$CuratedExhGenes < lowExhPct] <-
  "CD8_Res"


cell_attrsCD8_Tissue$TissueExhCells[is.na(cell_attrsCD8_Tissue$TissueExhCells)] <- "Other"

table(cell_attrsCD8_Tissue$TissueExhCells)
## save cell names for these groups:

CD8_Exh_Cells <- as.character(
  cell_attrsCD8_Tissue$CellName[
    cell_attrsCD8_Tissue$TissueExhCells == "CD8_Exh"])


CD8_Res_Cells <- as.character(
  cell_attrsCD8_Tissue$CellName[
    cell_attrsCD8_Tissue$TissueExhCells == "CD8_Res"])
```

Subset the Seurat object to also remove peripheral blood cells.
```{r}
CD8Tissue <- cd8[, !cd8$Tissue == "P" &
                   !cd8$Sub_Cluster %in% c("CD8-GPR183",
                                       "CD8-LEF1")]

# all(cell_attrsCD8_Tissue$CellName == colnames(CD8Tissue))

CD8Tissue@meta.data <- cell_attrsCD8_Tissue

CD8Tissue <- SetIdent(CD8Tissue, value = "TissueExhCells")
```

### CD8 Exh markers
Perform DE analysis through two approaches and add them up.
```{r,cache=T}
##------- Find markers for CD8-Exh vs CD8-TRM-nonEx
markers_CD8Exh <-
  FindMarkers(
    CD8Tissue,
    ident.1 = "CD8_Exh",
    ident.2 = "CD8_Res",
    only.pos = TRUE,
    min.pct = 0.3,
    logfc.threshold = 0.3
  )

markers_CD8Exh$gene <- rownames(markers_CD8Exh)
markers_CD8Exh$Marker <- "CD8_Exh"

## 322 genes
DEGs_CD8Exh <- markers_CD8Exh[markers_CD8Exh$avg_logFC > 0.3 & markers_CD8Exh$p_val_adj < 0.05, ]

# dim(DEGs_CD8Exh)
##----------- DE using MAST
# BiocManager::install("MAST")
# library(MAST)
markers_CD8Exh_mast <-
  FindMarkers(
    CD8Tissue,
    ident.1 = "CD8_Exh",
    ident.2 = "CD8_Res",
    only.pos = TRUE,
    min.pct = 0.3,
    logfc.threshold = 0.3, 
    test.use = "MAST"
  )

markers_CD8Exh_mast$gene <- rownames(markers_CD8Exh_mast)
markers_CD8Exh_mast$Marker <- "CD8_Exh"

## 302 genes
DEGs_CD8Exh_mast <- markers_CD8Exh_mast[markers_CD8Exh_mast$avg_logFC > 0.3 & markers_CD8Exh_mast$p_val_adj < 0.05, ]

# dim(DEGs_CD8Exh_mast)

## 287 common
# length(intersect(DEGs_CD8Exh_mast$gene, DEGs_CD8Exh$gene))

## --- Only add genes that came up in MAST and do not exist in 
## Wilcox test

DEGs_CD8Exh_mast <- DEGs_CD8Exh_mast[! rownames(DEGs_CD8Exh_mast) %in% rownames(DEGs_CD8Exh), ]

##--- 
DEGs_CD8Exh <- rbind(DEGs_CD8Exh, DEGs_CD8Exh_mast)


## Add GPCRs:
DEGs_CD8Exh$IsGPCR <- FALSE
DEGs_CD8Exh$IsGPCR[DEGs_CD8Exh$gene %in% gpr] <- TRUE
## Add TFs:
DEGs_CD8Exh$IsTF <- FALSE
DEGs_CD8Exh$IsTF[DEGs_CD8Exh$gene %in% TFs$Gene] <- TRUE

```


#### Percentile filtration
Filter those high in blood cells and also high in Res cells.
We keep genes whose 65% tile expression in Exh cells are larger than the 90%tile of their expression in blood and Res cells. This results in 27 genes.

Only two genes (CXCL13 and HAVCR2): their 25%tile > others 90%tile
```{r}
exprData <- as.matrix(GetAssayData(cd8, slot = "data"))

gg <-  DEGs_CD8Exh$gene
exprData_Genes <- exprData[gg, ]
cell_attrsCD8 <- cd8[[]]


Exh_Pct <-
  rowQuantiles(exprData_Genes[, CD8_Exh_Cells], probs = 0.65)

Res_Pct <-
  rowQuantiles(exprData_Genes[, CD8_Res_Cells], probs = 0.9)

Blood_Pct <-
  rowQuantiles(exprData_Genes[, cell_attrsCD8$Tissue == "P"], probs = 0.9)


pctData_cd8Exh <- data.frame(cbind(Exh_Pct,
                               Res_Pct, 
                               Blood_Pct
                               ))

pctData_cd8Exh$gene <- rownames(pctData_cd8Exh)

pctData_cd8ExhPass <- pctData_cd8Exh %>% 
  filter(Exh_Pct > Res_Pct & Exh_Pct> Blood_Pct) %>% 
  data.frame()

pctData_cd8ExhPass <-
  pctData_cd8ExhPass[order(pctData_cd8ExhPass$Exh_Pct, decreasing = T),]

DT::datatable(pctData_cd8ExhPass, filter = "top") ## 27 genes


DEGs_CD8Exh$PctPass <- FALSE
DEGs_CD8Exh$PctPass[DEGs_CD8Exh$gene %in% pctData_cd8ExhPass$gene] <- TRUE

# write.table(
#   DEGs_CD8Exh,
#   paste0(outPath, "Zhang_CD8Exh_Markers_BasedOnScores_20200928.txt"),
#   row.names = F,
#   sep = "\t"
# )
```


### CD8 Res markers
```{r}
##------- Find markers for CD8-TRM vs CD8-Exh
markers_CD8Res <-
  FindMarkers(
    CD8Tissue,
    ident.1 = "CD8_Res",
    ident.2 = "CD8_Exh",
    only.pos = TRUE,
    min.pct = 0.3,
    logfc.threshold = 0.3
  )

markers_CD8Res$gene <- rownames(markers_CD8Res)
markers_CD8Res$Marker <- "CD8_Res"

## 207
DEGs_CD8Res <- markers_CD8Res[markers_CD8Res$avg_logFC > 0.3 & markers_CD8Res$p_val_adj < 0.05, ]

# dim(DEGs_CD8Res)
##------- use MAST for DE
markers_CD8Res_mast <-
  FindMarkers(
    CD8Tissue,
    ident.1 = "CD8_Res",
    ident.2 = "CD8_Exh",
    only.pos = TRUE,
    min.pct = 0.3,
    logfc.threshold = 0.3,
    test.use = "MAST"
  )

markers_CD8Res_mast$gene <- rownames(markers_CD8Res_mast)
markers_CD8Res_mast$Marker <- "CD8_Res"

## 192 genes

DEGs_CD8Res_mast <- markers_CD8Res_mast[markers_CD8Res_mast$avg_logFC > 0.3 & markers_CD8Res_mast$p_val_adj < 0.05, ]

# dim(DEGs_CD8Res_mast)
## 184 common
# length(intersect(DEGs_CD8Res_mast$gene, DEGs_CD8Res$gene))

##  23 genes came up in this method that did not come up in Wilcox!
DEGs_CD8Res_mast <- DEGs_CD8Res_mast[! rownames(DEGs_CD8Res_mast) %in% rownames(DEGs_CD8Res), ]

##---
DEGs_CD8Res <- rbind(DEGs_CD8Res, DEGs_CD8Res_mast)
# dim(DEGs_CD8Res)

## Add GPCRs:
DEGs_CD8Res$IsGPCR <- FALSE
DEGs_CD8Res$IsGPCR[DEGs_CD8Res$gene %in% gpr] <- TRUE
## Add TFs:
DEGs_CD8Res$IsTF <- FALSE
DEGs_CD8Res$IsTF[DEGs_CD8Res$gene %in% TFs$Gene] <- TRUE

```


#### Percentile filtration
Filter those high in blood and also high in Exh cells --> 43 Res genes
```{r}
exprData <- as.matrix(GetAssayData(cd8, slot = "data"))

gg <-  DEGs_CD8Res$gene
exprData_Genes <- exprData[gg, ]

cell_attrsCD8 <- cd8[[]]


Res_Pct <-
  rowQuantiles(exprData_Genes[, CD8_Res_Cells ], probs = 0.65)

Exh_Pct <-
  rowQuantiles(exprData_Genes[,CD8_Exh_Cells ], probs = 0.9)

Blood_Pct <-
  rowQuantiles(exprData_Genes[, cell_attrsCD8$Tissue == "P"], probs = 0.90)


pctData_cd8Trm <- data.frame(cbind(Res_Pct,
                               Exh_Pct, 
                               Blood_Pct
                               ))

pctData_cd8Trm$gene <- rownames(pctData_cd8Trm)

pctData_cd8ResPass <- pctData_cd8Trm %>% 
  filter(Res_Pct > Exh_Pct & Res_Pct > Blood_Pct) %>% 
  data.frame()

pctData_cd8ResPass <-
  pctData_cd8ResPass[order(pctData_cd8ResPass$Res_Pct, decreasing = T),]

DT::datatable(pctData_cd8ResPass, filter = "top")  ## 43

DEGs_CD8Res$PctPass <- FALSE
DEGs_CD8Res$PctPass[DEGs_CD8Res$gene %in% pctData_cd8ResPass$gene] <- TRUE

# write.table(
#   DEGs_CD8Res,
#   paste0(outPath, "Zhang_CD8Res_Markers_BasedOnScores_20200928.txt"),
#   row.names = F,
#   sep = "\t"
# )

```


## Score SMART-seq2 CD8 cells using the final Res and Exh signatures
```{r}
library(singscore)
exprData <- as.matrix(GetAssayData(cd8, slot = "data"))
rd <- rankGenes(exprData)

CD8_Res <- simpleScore(
  rankData = rd,
  upSet =  pctData_cd8ResPass$gene,
  knownDirection = T,
  centerScore = F
)

CD8_Exh <- simpleScore(
  rankData = rd,
  upSet =  pctData_cd8ExhPass$gene,
  knownDirection = T,
  centerScore = F
)

cd8$CD8_Res <- CD8_Res$TotalScore
cd8$CD8_Exh <- CD8_Exh$TotalScore


## ---- UMAP coloured by the two scores

# cell_attrsCD8 <- cd8[[]]
# currentRedData <- cd8@reductions[["umap"]]@cell.embeddings
# 
# cell_attrsCD8$UMAP_1 <- currentRedData[, "UMAP_1"]
# cell_attrsCD8$UMAP_2 <- currentRedData[, "UMAP_2"]


# for(i in c("CD8_Res", "CD8_Exh")){
#   print(
#     cell_attrsCD8 %>% 
#       arrange(!!sym(i)) %>% 
#     ggplot(., 
#            aes_string(x = "UMAP_1", 
#                    y = "UMAP_2",
#                    color = i)) +
#   geom_point(alpha = 0.8, cex = 0.5) +
#   scale_color_viridis_c() +
#   theme_dark() +
#           theme(
#         legend.spacing = unit(1.5, "mm"),
#         legend.position = "top",
#         legend.key.width = unit(0.9, "cm"), 
#         legend.title = element_text(size = 8, face = "italic"),
#         legend.text = element_text(size = 8)
#       )
#   )
# }
```

### Save data after final scores
```{r}
# saveRDS(cd8, paste0(outPath, "DontPush/Zhang_SmartSeq_SeuratObj_CD8_20200928_UMAP_scores2.RDS"))
# cd8 <- readRDS(paste0(outPath, "DontPush/Zhang_SmartSeq_SeuratObj_CD8_20200928_UMAP_scores2.RDS"))
```


## PCA/UMAP using marker genes
```{r, warning = F, message=F, cache = T}
gg <-  c(pctData_cd8ExhPass$gene, pctData_cd8ResPass$gene)

cd8TrmExhSig <-
  RunPCA(cd8, features = rownames(cd8)[rownames(cd8) %in% gg], verbose = F)

# ElbowPlot(cd8TrmExhSig)

## Clustering and UMAP on marker genes - including both signatures
cd8TrmExhSig <- FindNeighbors(cd8TrmExhSig, reduction = "pca", dims = 1:10)
cd8TrmExhSig <- FindClusters(cd8TrmExhSig, resolution = 0.5)

cd8TrmExhSig <- RunUMAP(cd8TrmExhSig, reduction = "pca", 
          dims = 1:10,  min.dist = 0.3, n.neighbors = 30)

# DimPlot(cd8TrmExhSig, reduction = "umap")
# DimPlot(cd8TrmExhSig, reduction = "umap", group.by = "Sub_Cluster")
# DimPlot(cd8TrmExhSig, reduction = "umap", group.by = "Tissue")
# FeaturePlot(cd8TrmExhSig, reduction = "umap", features = "CD8_Exh")
# FeaturePlot(cd8TrmExhSig, reduction = "umap", features = "CD8_Res")
```


## Trajectory analysis

```{r}
## add clusters from markers:
cd8$seurat_clusters_markers <- cd8TrmExhSig$seurat_clusters

cell_attrsCD8 <- cd8[[]]
exprData <- as.matrix(GetAssayData(cd8, slot = "data"))

rd_umap_seurat <- cd8@reductions[["umap"]]@cell.embeddings
rd_pca_seurat <- cd8@reductions[["pca"]]@cell.embeddings
```


### Generate SCE
```{r}
cd8.sce <- as.SingleCellExperiment(cd8)

reducedDims(cd8.sce) <-
  SimpleList(
    ## from Seurat
    PCA_Seurat = rd_pca_seurat,
    UMAP_Seurat = rd_umap_seurat,
    
    ## scores
    CuratedScores = as.matrix(cell_attrsCD8[, c("CuratedTrmGenes", "CuratedExhGenes")]),
    MarkerScores = as.matrix(cell_attrsCD8[, c("CD8_Res", "CD8_Exh")]),
    # FinalScores = as.matrix(cell_attrsCD8[, c("Trm", "Exh")]),
    
    ## PCs from markers
    UMAP_markers = cd8TrmExhSig@reductions[["umap"]]@cell.embeddings,
    PCA_markers = cd8TrmExhSig@reductions[["pca"]]@cell.embeddings
  )


# saveRDS(cd8.sce, paste0(outPath, "Zhang_SmartSeq_CD8_SingleCellExperiment_20201016.RDS"))
# cd8.sce <- readRDS(paste0(outPath, "Zhang_SmartSeq_CD8_SingleCellExperiment_20201016.RDS"))
```


### Slingshot (realted to Figure 1 in the paper)

```{r, fig.cap= "Trajectory analysis in CD8 T cells using the UMAP and clusters obtained based on our Res and Exh marker genes, coloured by Res and Exh signature scores"}
source(paste0(scriptPath, "slingshot_ggplot.R"))

cd8.sce <-
  slingshot(cd8.sce, clusterLabels = 'seurat_clusters_markers', reducedDim = 'UMAP_markers')

sds <- SlingshotDataSet(cd8.sce)

textSize <- 1.4

currentThemeTrj <- theme_dark() +
  theme(
    # panel.background = element_blank()
    axis.title = element_text(size = rel(textSize)),
    axis.text = element_text(angle = 0, size = rel(textSize)),
    # strip.background = element_rect(colour = "#f0f0f0", fill = "#f0f0f0"),
    strip.text = element_text(size = rel(textSize)),
    axis.line = element_line(colour = "black", size = 0.5),
    legend.position = 'top',
    legend.title = element_text(size = rel(textSize), face = "italic"),
    legend.text = element_text(size = rel(textSize)),
     legend.key.width = unit(2.4, 'lines'),
    ## increase the line space
    plot.title = element_text(
      face = "bold",
      size = rel(textSize),
      hjust = 0.5
    )
  )


# pdf(
#   paste0(
#     figPath,
#     "Trajectory_ggplot_ZhangSmart_CD8_MarkersUMAP_ClusterMarkers_ExhScore.pdf"
#   ),
#   height = 5,
#   width = 8
# )
ggplot() +
  geom_point(size = 1) +
  gg_plot(sds, col = cd8.sce$CD8_Exh) +
  xlab("UMAP 1") +
  ylab("UMAP 2") +
  scale_color_viridis_c(
    option = "viridis", 
    name = "CD8 \nexhaustion\nscore")  +
  currentThemeTrj 
# dev.off()

# pdf(
#   paste0(
#     figPath,
#     "Trajectory_ggplot_ZhangSmart_CD8_MarkersUMAP_ClusterMarkers_ResScore.pdf"
#   ),
#   height = 5,
#   width = 8
# )
ggplot() +
  geom_point(size = 1) +
  gg_plot(sds, col = cd8.sce$CD8_Res) +
  xlab("UMAP 1") +
  ylab("UMAP 2") +
  scale_color_viridis_c(
    option = "viridis", 
    name = "CD8 \nresidency\nscore")  +
  currentThemeTrj 
# dev.off()
```

Check the expression of some selected genes on the trajectory
```{r, fig.cap="Trajectory analysis in CD8 T cells using the UMAP and clusters obtained based on our Res and Exh marker genes, coloured by the expression of selected genes"}

gs <-
  c(
    "IFNG",
    "GZMK",
    "CCR7",
    "SELL",
    "GPR183",
    "GPR15",
    "PDCD1",
    "CTLA4",
    "HAVCR2",
    "TIGIT"
  )

# pdf(
#   paste0(
#     figPath,
#     "Trajectory_ggplot_ZhangSmart_CD8_MarkersUMAP_ClusterMarkers_selGenes.pdf"
#   ),
#   height = 5,
#   width = 8
# )
for(i in gs){
  d0 <- as.matrix(assay(cd8.sce))[i,]

print(ggplot() +
  geom_point(size = 1) +
  gg_plot(sds, col = d0) +
  # gg_plot(sds, col = nk.sce$NK_Exh) +
  xlab("UMAP 1") +
  ylab("UMAP 2") +
  scale_color_viridis_c(
    option = "viridis", 
    name = i)  +
    # name = "NK \nexhaustion\nscore")  +
  currentThemeTrj )
}
# dev.off()
```

# Filter markers based on the tumour cell data
## CCLE

```{r, cache = T, warning=F, message=F, fig.cap="Expression of CD8 Res and Exh genes in the CCLE  CRC data", fig.height=12, fig.width=8 }
selectedGenes <- c(pctData_cd8ResPass$gene, pctData_cd8ExhPass$gene)

dge <- readRDS(paste0(ccleLcmPath, "DGEList_CCLE_panCancer.RDS"))

library(edgeR)
  
kp <- rowSums(cpm(dge) > 1) >= 5 |
  dge$genes$Symbol %in% selectedGenes
summary(kp)

dgeFilt <- dge[kp, ]

dgeFiltNorm <- calcNormFactors(dgeFilt)

logRPKM <- rpkm(dgeFiltNorm, log = T, gene.length = dgeFiltNorm$genes$gene_length)
gg <- selectedGenes[selectedGenes %in% dgeFiltNorm$genes$Symbol]

logRPKM_CRC <- logRPKM[, dgeFiltNorm$samples$type_refined == "colorectal"]

brbg <- rev(RColorBrewer::brewer.pal(11, "PRGn"))
brbg_hcl <- colorspace::diverging_hcl(11,
  h = c(180, 50), c = 80, l = c(20, 95), power = c(0.7, 1.3))

ComplexHeatmap::Heatmap(logRPKM_CRC[gg, ! is.na(colnames(logRPKM_CRC))], 
                        col = brbg_hcl, 
                        show_column_names = F, name = "logRPKM")


```

We subset genes to those that have low expression in CCLE. These are genes whose median expression is =< 0 across samples.
```{r}
ggLogRPKM <- logRPKM_CRC[gg, ! is.na(colnames(logRPKM_CRC))]
markersPassCCLE <- rownames(ggLogRPKM)[rowMedians(ggLogRPKM) <= 0]
markersPassCCLE
```


## LCM (Laser captured micro-dissection)
Microarray data downloaded and processed from GSE21510.
```{r, cache = T, warning=F, message=F, fig.cap="Expression of CD8 Res and Exh genes in the LCM  CRC data", fig.height=12, fig.width=8}
ll <- readRDS(paste0(ccleLcmPath, "ColoRectal_GeneExpression_LCM_MicroArray_GSE21510_AvergedProbes_Filtered.rds"))

## 67/70 genes --> missing:  "KRT86"    "APOBEC3H" "TTC24" 
gg <- selectedGenes[selectedGenes %in% rowData(ll)$Gene.Symbol]

l <- ll[rowData(ll)$Gene.Symbol %in% gg , ll$`tissue:ch1` == "cancer, LCM" ]

lAnnot <- colData(l)

# lAnnot$Stage[grepl("1", lAnnot$`stage:ch1`)] <- "I"
# lAnnot$Stage[grepl("2", lAnnot$`stage:ch1`)] <- "II"
# lAnnot$Stage[grepl("3", lAnnot$`stage:ch1`)] <- "III"
# lAnnot$Stage[grepl("4", lAnnot$`stage:ch1`)] <- "IV"

lExpr <- assay(l)
row.names(lExpr) <- rowData(l)$Gene.Symbol

brbg <- rev(RColorBrewer::brewer.pal(11, "PRGn"))
brbg_hcl <- colorspace::diverging_hcl(11,
  h = c(180, 50), c = 80, l = c(20, 95), power = c(0.7, 1.3))

ComplexHeatmap::Heatmap(lExpr, 
                        col = brbg_hcl, 
                        show_column_names = F, name = "logExpr")

```
We also keep genes whose median expression is =< median of all genes in the data.
```{r}
th <- quantile(rowMedians(assay(ll)), probs = 0.5)
markersPassLCM <- rownames(lExpr)[rowMedians(lExpr) <= th]
markersPassLCM
# [1] "HAVCR2"   "PDE4A"    "GFPT2"    "CXCL13"   "XCL1"    
#  [6] "IL12RB2"  "CSF1"     "TNFSF4"   "SARDH"    "NR4A3"   
# [11] "MYO7A"    "GPR15"    "APOBEC3C" "GPR65"    "CTLA4"   
# [16] "AFAP1L2"

# intersect(markersPassLCM, markersPassCCLE)
#  [1] "HAVCR2"  "GFPT2"   "CXCL13"  "XCL1"    "IL12RB2" "CSF1"   
#  [7] "TNFSF4"  "SARDH"   "NR4A3"   "MYO7A"   "GPR15"   "GPR65"  
# [13] "CTLA4" 

# markersPassLCM[! markersPassLCM %in% markersPassCCLE]
# "PDE4A"    "APOBEC3C" "AFAP1L2" 
```


All Res and Exh genes that passed the threshold in either of the two data sets:
```{r}
passGenes <- unique(c(markersPassLCM, markersPassCCLE))
passGenes
```


Residency genes that pass the thresholds in either of the two data sets:
```{r}
ResPass <- passGenes[ passGenes %in% pctData_cd8ResPass$gene]
ResPass
```

Exhaustion genes that pass the thresholds in either of the two data sets:
```{r}
ExhPass <- passGenes[ passGenes %in% pctData_cd8ExhPass$gene]
ExhPass
```

Add these information to the signatures and export them.
```{r}

resData <- pctData_cd8ResPass

resData$CLsPass <- FALSE
resData$CLsPass[resData$gene %in% markersPassCCLE] <- TRUE

resData$LcmPass <- FALSE
resData$LcmPass[resData$gene %in% markersPassLCM] <- TRUE

resData$CancerPass <- FALSE
resData$CancerPass[resData$gene %in% ResPass] <- TRUE



exhData <- pctData_cd8ExhPass

exhData$CLsPass <- FALSE
exhData$CLsPass[exhData$gene %in% markersPassCCLE] <- TRUE

exhData$LcmPass <- FALSE
exhData$LcmPass[exhData$gene %in% markersPassLCM] <- TRUE

exhData$CancerPass <- FALSE
exhData$CancerPass[exhData$gene %in% ExhPass] <- TRUE

```


### Save merged markers
```{r}
exhData$Marker <- "Exh"
resData$Marker <- "Res"

# ResExhDataList <- list(exhData = exhData, 
#                        resData = resData)
# saveRDS(ResExhDataList, paste0(outPath, "ResExhMarkers_CD8_CLs_LCM.RDS"))

ResExhData <- rbind(resData, exhData)

write.table(
  ResExhData,
  paste0(outPath, "ResExhMarkers_CD8_CLs_LCM.txt"),
  row.names = F,
  sep = "\t"
)
```



# Session information
Here are the list of tools and packages (along with their versions) used in this document. 
```{r}
sessionInfo()
```

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 598px;"></div>




