---
title: "Marisa COAD survival"
output: html_document
---

# Set up
```{r}
mainDir <- getwd()
outPath <- "../output/Microarray/paper/"
figPath <- "../figure/Microarray/paper/"
dataPath <- "../data/"
scriptPath <- "../script/"

crcPath <- "/Users/mfor0011/Documents/data/CRC/CRC_GEO/"
tpmPath <- "/Users/mfor0011/Documents/data/NK_Data/singleCell_NK/Human/GSE146771_Zhang_CRC/"
sigPath <- "/Users/mfor0011/Documents/data/signatures/ProcessedSigs/"

ifelse(!dir.exists(file.path(mainDir, outPath)), dir.create(file.path(mainDir, outPath)), FALSE)
ifelse(!dir.exists(file.path(mainDir, figPath)), dir.create(file.path(mainDir, figPath)), FALSE)
ifelse(!dir.exists(file.path(mainDir, dataPath)), dir.create(file.path(mainDir, dataPath)), FALSE)
ifelse(!dir.exists(file.path(mainDir, scriptPath)), dir.create(file.path(mainDir, scriptPath)), FALSE)

# library(ggplot2)
library(tidyverse)
library(RColorBrewer)
library(ggbeeswarm)
library(SingleCellExperiment)
library(Seurat)
library(hgu133a.db)

options(digits = 3)

equal_breaks <- function(n = nBreak, s = scalingFactor, ...){
  function(x){
    # rescaling
    d <- s * diff(range(x)) / (1+2*s)
    round( seq(min(x)+d, max(x)-d, length=n), 2)
  }
}

nBreak = 3
scalingFactor = 0.05

textSize <- 1.2

currentTheme <- theme_bw() +
  theme(
    # panel.background = element_blank()
    axis.title = element_text(size = rel(textSize)),
    axis.text = element_text(angle = 0, size = rel(textSize)),
    # strip.background = element_rect(colour = "#f0f0f0", fill = "#f0f0f0"),
    strip.text = element_text(size = rel(textSize)),
    axis.line = element_line(colour = "black", size = 0.5),
    legend.position = 'top',
    legend.title = element_text(size = rel(textSize), face = "italic"),
    legend.text = element_text(size = rel(textSize)),
    legend.key.size = unit(1, 'lines'),
    ## increase the line space
    plot.title = element_text(
      face = "plain",
      size = rel(textSize),
      hjust = 0.5
    )
  )

cols <-  c(
  brewer.pal(8, "Dark2")[-5],
  brewer.pal(10, "Paired"),
  brewer.pal(12, "Set3"),
  brewer.pal(9, "Blues")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Oranges")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Greens")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Purples")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Reds")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Greys")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "BuGn")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "PuRd")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "BuPu")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "YlGn")[c(8, 3, 7, 4, 6, 9, 5)])


source(paste0(scriptPath, "visReduction_function.R"))
source("../script/coxph_analysis.R")
source("../script/survival_plot.R")
```
# Read signatures
```{r}
source(paste0(scriptPath, "ReadSignaturesTum_paper.R"))
```

# Marisa data
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3660251/

```{r}
library(hgu133plus2.db)
library(SummarizedExperiment)
# mr <- readRDS(paste0(crcPath, "GSE39582/GSE39582_SummExpr.RDS"))
# assay(mr)[1:4,1:5]
# 
# mr <- mr[! grepl("///", rowData(mr)$Gene.Symbol), ]
# 
# mrExpr <- data.frame(as.matrix(assay(mr)))
# mrAnn <- rowData(mr)
# mrAnn$Gene.Symbol <- as.character(mrAnn$Gene.Symbol)
# mrExpr$Symbol <- mrAnn$Gene.Symbol
# 
# eMr <- mrExpr %>% 
#   group_by(Symbol) %>% 
#   summarise_all(median)
# 
# eMar <- as.matrix(eMr[-1, 2:ncol(eMr)])
# rownames(eMar) <- eMr$Symbol[-1]
# 
# ###---- annotaion:
# annMar <- colData(mr)
# annMar$source_name_ch1 <- as.character(annMar$source_name_ch1)
# annMar$Sample_Type[grepl("non tumoral", annMar$source_name_ch1)] <- "Normal"
# annMar$Sample_Type[grepl("primary colorectal Adenocarcinoma", annMar$source_name_ch1)] <- "Cancer"
# 
# MarisaDataList <- list(eMar = eMar, annMar = annMar)

# saveRDS(MarisaDataList, paste0(outPath, "Marisa_Expr_Annot.RDS"))

MarisaDataList <- readRDS(paste0(outPath, "Marisa_Expr_Annot.RDS"))
eMar <- MarisaDataList$eMar
annMar <- MarisaDataList$annMar

all(rownames(annMar) == colnames(eMar))


colnames(annMar)[colnames(annMar) == "braf.mutation.ch1"] <- "BRAF"
colnames(annMar)[colnames(annMar) == "kras.mutation.ch1"] <- "KRAS"
colnames(annMar)[colnames(annMar) == "tp53.mutation.ch1"] <- "TP53"

colnames(annMar)[colnames(annMar) == "tnm.stage.ch1"] <- "Stage"
annMar$Stage[annMar$Sample_Type == "Normal"] <- "N"
annMar$Stage[annMar$Stage == 0] <- NA

annMar$Stage[annMar$Stage == 1] <- "I"
annMar$Stage[annMar$Stage == 2] <- "II"
annMar$Stage[annMar$Stage == 3] <- "III"
annMar$Stage[annMar$Stage == 4] <- "IV"


##----- filter for Cancer samples only:
eMarFilt <- eMar[, annMar$Sample_Type == "Cancer"]
annMarFilt <- annMar[ annMar$Sample_Type == "Cancer", ]

### annMarFilt <- annMarFilt[annMarFilt$tnm.stage.ch1 != "0", ]

annMarFilt$os.delay.ch1 <- as.numeric(annMarFilt$os.delay.ch1)

colnames(annMarFilt)[colnames(annMarFilt) == "os.delay.ch1"] <- "OS.time"
colnames(annMarFilt)[colnames(annMarFilt) == "os.event.ch1"] <- "OS"
colnames(annMarFilt)[colnames(annMarFilt) == "rfs.delay.ch1"] <- "RFS.time"
colnames(annMarFilt)[colnames(annMarFilt) == "rfs.event.ch1"] <- "RFS"

colnames(annMarFilt)[colnames(annMarFilt) == "mmr.status.ch1"] <- "MMR"
colnames(annMarFilt)[colnames(annMarFilt) == "cin.status.ch1"] <- "CIN"
colnames(annMarFilt)[colnames(annMarFilt) == "cimp.status.ch1"] <- "CIMP"
colnames(annMarFilt)[colnames(annMarFilt) == "age.at.diagnosis..year..ch1"] <- "Age"


annMarFilt$Age <- as.numeric(annMarFilt$Age)
annMarFilt$OS <- as.numeric(annMarFilt$OS)
annMarFilt$OS.time <- as.numeric(annMarFilt$OS.time)
annMarFilt$RFS <- as.numeric(annMarFilt$RFS)
annMarFilt$RFS.time <- as.numeric(annMarFilt$RFS.time)

annMarFilt$Age_Status[annMarFilt$Age > median(annMarFilt$Age, na.rm = T)] <- "Old"
annMarFilt$Age_Status[annMarFilt$Age <= median(annMarFilt$Age, na.rm = T)] <- "Young"

```

## CMS
```{r}
# eMarFilt <- eMar[, annMar$Sample_Type == "Cancer"]
# annMarFilt <- annMar[ annMar$Sample_Type == "Cancer", ]

set.seed(20201011)

cmsMar <-
  CMScaller::CMScaller(
    emat = eMarFilt,
    RNAseq = F,
    FDR = 0.05,
    rowNames = "symbol"
  )
# CMS1 CMS2 CMS3 CMS4 <NA> 
#   84  175   83  154   70 

annMarFilt$CMS <- as.character(cmsMar$prediction)

## add CMS to the annotation data including normal samples
annMar <-
  merge(
    annMar,
    cmsMar[, "prediction", drop = F],
    by.x = "geo_accession",
    by.y = "row.names",
    all.x = T
  )

colnames(annMar)[colnames(annMar) == "prediction"] <- "CMS"
annMar$CMS <- as.character(annMar$CMS)

annMar$CMS[annMar$Sample_Type == "Normal"] <- "N"

row.names(annMar) <- annMar$geo_accession
annMar <- annMar[colnames(eMar), ]

all(rownames(annMar) == colnames(eMar))


```

## Score samples



```{r}
# gg <- unique(
#   c(
#     allMarkers$gene[allMarkers$CancerPass],
#     "TGFB1", "TGFB2", "TGFB3", "TGFBR1", "TGFBR2",
#     "IL15", "IL12A", "IL12B", "IL18", "IL2",
#     "IFNG", "HIF1A", "FAS", "FASLG",
#     # Hypoxia_Buffa,
#     intersect(checkSigID$Symbol[checkSigID$geneset == "HALLMARK_HYPOXIA"],
#               Hypoxia_Buffa),
#     intersect(checkSigID$Symbol[checkSigID$geneset == "HALLMARK_HYPOXIA"],
#               sigsTum$Hypoxia),
#     CanEvasion 
#   )
# )

kpGenes <- readRDS("../kpGenes.RDS")

kp1 <- rowSums(eMar > 4) >= 0.5 * ncol(eMar)
kp2 <- rownames(eMar)  %in% unique(
  c(
    kpGenes
    # gg
  )
)


eMar <- eMar[ kp1 | kp2, ]


rdMar <- rankGenes(eMar)

marScoresDisp <- multiScore(rdMar, 
                         upSetColc = AllSigsCollection, 
                         knownDirection = T, 
                         centerScore = F)

marScores <-
  data.frame(t(marScoresDisp$Scores),
             check.rows = F,
             check.names = F)

## Add TGFb-EMT bidirectional scores too:
marScores$TGFbEMT_Dir <- simpleScore(
  rdMar,
  upSet = sigsTum$TGFb_EMT$tgfb_Up,
  downSet = sigsTum$TGFb_EMT$tgfb_Dn,
  knownDirection = T,
  centerScore = F
)$TotalScore

scoreNames <- c(colnames(marScores))
scoreNames <- scoreNames[order(scoreNames)]


##----- Add these to annotation data:
annMar <-
  data.frame(annMar ,
             marScores,
             check.names = F,
             check.rows = F)


annMarFilt <- data.frame(annMarFilt,
                          annMar[row.names(annMarFilt), scoreNames], 
                          check.names = F)


```


## Scores vs Annotation
### Define Cancer and Immune signatures
```{r}
canSigs <- scoreNames[! grepl("Res", scoreNames) &
                        ! grepl("Exh", scoreNames)]

imSigs <- scoreNames[grepl("Res", scoreNames) |
                        grepl("Exh", scoreNames)]
 
```

### General annotations: Stage, subtypes, gender, location, chemo
```{r}
library(ggsignif)

scoreColumns <- which(colnames(annMar)  %in% scoreNames)

annMarLong <- annMar %>%
  pivot_longer(
    cols = scoreColumns[1]:scoreColumns[length(scoreColumns)],
    names_to = "Signature",
    values_to = "Score"
  ) %>%
  data.frame(check.rows = F, check.names = F)

cmsCol <- c(
  brewer.pal(11, "BrBG")[3],
  brewer.pal(11, "PRGn")[2],
  brewer.pal(9, "Blues")[5],
  brewer.pal(11, "BrBG")[10],
  "grey50"
)

pdf(paste0(figPath, "Marisa_Boxplot_Scores_CMS_Bulk.pdf"), height = 6, width = 5.5)
annMarLong %>%
  filter(complete.cases(CMS)) %>%
  filter(grepl("Bulk", Signature)) %>%
  filter(! grepl("CD8NK", Signature) & ! grepl("All", Signature)) %>%
  ggplot(., aes(x = CMS, y = Score, color = CMS)) +
  geom_boxplot(outlier.shape = NA) +
  facet_wrap( ~ Signature, scales = "free", ncol = 2) +
  scale_y_continuous(breaks = equal_breaks(n = nBreak, s = scalingFactor),
                     expand = c(scalingFactor, 0)) +
  scale_color_manual(values = cmsCol) +
  currentTheme +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
        axis.ticks = element_blank(),
    legend.position = "bottom"
    )
dev.off()


annMarLong$tumor.location.ch1[
  annMarLong$tumor.location.ch1 == "distal"
] <- "Distal"

annMarLong$tumor.location.ch1[
  annMarLong$tumor.location.ch1 == "proximal"
] <- "Proximal"

pdf(
  paste0(figPath, "Marisa_Boxplot_Scores_Location_Cancer_Bulk.pdf"),
  height = 6,
  width = 5
)
annMarLong %>%
  filter(Sample_Type == "Cancer") %>%
  filter(grepl("Bulk", Signature)) %>%
  filter(!grepl("CD8NK", Signature) & !grepl("All", Signature)) %>%
  ggplot(.,
         aes(x = tumor.location.ch1, y = Score, color = tumor.location.ch1)) +
  geom_boxplot(outlier.shape = NA) +
  ggpubr::geom_signif(
    comparisons = list(c("Distal", "Proximal")),
    map_signif_level = TRUE,
    tip_length = 0,
    vjust = 1.4,
    col = "gray40"
  ) +
  labs(colour = "Location") +
  facet_wrap(~ Signature, scales = "free", ncol = 2) +
  scale_y_continuous(breaks = equal_breaks(n = nBreak, s = scalingFactor),
                     expand = c(scalingFactor, 0)) +
scale_color_manual(values = brewer.pal(8, "Greys")[c(5, 7)]) +
  currentTheme +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks = element_blank(),
    legend.position = "bottom"
  )
dev.off()

```

### MSI 

MSI was analyzed using a panel of five different microsatellite loci from the Bethesda reference panel. MSI-high tumors were further classified as deficient MMR (dMMR), and both MSI-low and MSS tumors as proficient MMR (pMMR). 

CIN (chromosomal instability) was defined from CGH profiles: samples with at least 20% gain or loss of whole chromosomes or fractions of chromosomes were scored as CIN+ 

CIMP (CpG island methylator phenotype) status was determined using a panel of five markers (CACNA1G, IGF2, NEUROG1, RUNX3, and SOCS1) 

```{r}

colnames(annMarLong)[colnames(annMarLong) == "mmr.status.ch1"] <- "MMR"


pdf(paste0(figPath, "Marisa_Boxplot_Scores_MMR_Bulk.pdf"), height = 6, width = 5.2)

annMarLong %>%
  filter(Sample_Type == "Cancer") %>%
  filter(MMR != "N/A") %>%
  filter(grepl("Bulk", Signature)) %>%
  filter(! grepl("CD8NK", Signature) & ! grepl("All", Signature)) %>%
  ggplot(., aes(x = MMR, y = Score, color = MMR)) +
  geom_boxplot() +
  geom_signif(
    comparisons = list(c("dMMR", "pMMR")),
    map_signif_level = TRUE,
    tip_length = 0,
    vjust = 1.4,
    col = "gray40"
  ) +
  facet_wrap( ~ Signature, scales = "free", ncol = 2) +
  scale_y_continuous(breaks = equal_breaks(n = nBreak, s = scalingFactor),
                     expand = c(scalingFactor, 0)) +
  scale_color_manual(values =  rev(c(brewer.pal(9, "Purples") [c(4, 6, 9)]))) +
  currentTheme +
    theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
        axis.ticks = element_blank(),
    legend.position = "bottom"
    )
dev.off()

```

### Scatterplot of score pairs
```{r}
colnames(annMar)[colnames(annMar) == "mmr.status.ch1"] <- "MMR"  
annMar$MMR <- as.character(annMar$MMR)
annMar$MMR[annMar$MMR  %in% "N/A"] <- NA

cmsCol <- c(
  brewer.pal(11, "BrBG")[3],
  brewer.pal(11, "PRGn")[2],
  brewer.pal(9, "Blues")[5],
  brewer.pal(11, "BrBG")[10]
)


pdf(
  paste0(
    figPath,
    # "ScatterPlots_Marisa_ExhTrmCancerScorePairs_colMMR_rmNormal_Lines.pdf"
    "ScatterPlots_Marisa_ExhTrmCancerScorePairs_colCMS_rmNormal_Lines.pdf"
  ),
  height = 4.7,
  width = 4.3
)
for(im in imSigs) {
  for (ca in canSigs) {
    
      palette <- cmsCol
      # palette <- rev(c(brewer.pal(9, "Purples") [c( 6, 9)]))
      pal <- list(values=palette, na.value="gray80")
    
      d <- annMar %>%
      filter(! annMar$Sample_Type  %in%  "Normal") %>%
      mutate(medIm = median(!!sym(im)),
             medCa = median(!!sym(ca)))
      
          corVal <- round(cor.test(d[, im], d[, ca])[[4]], 2)
          
    print(
      d %>% 
        ggplot(.,
              # aes_string(im, ca, color = "MMR")) +
               aes_string(im, ca, color = "CMS")) +
        geom_point(alpha = 0.8) +
        geom_vline(xintercept = d$medIm) +
        geom_hline(yintercept = d$medCa) +
        invoke(scale_colour_manual, pal) +
        scale_x_continuous(
                       breaks = equal_breaks(n = nBreak, s = scalingFactor),
                       expand = c(scalingFactor, 0)
                     ) +
        scale_y_continuous(
                       breaks = equal_breaks(n = nBreak, s = scalingFactor),
                       expand = c(scalingFactor, 0)
                     ) +
        ggtitle(paste0("Marisa (cor = ", corVal, ")")) +
        currentTheme + 
        guides(color = guide_legend(nrow=2))
    )
  }
}

dev.off()



pdf(
  paste0(
    figPath,
    "ScatterPlots_Marisa_ExhTrmScorePairs_colCMS_rmNormal_Lines.pdf"
    # "ScatterPlots_Marisa_ExhTrmScorePairs_colMMR_rmNormal_Lines.pdf"
  ),
  height = 4.7,
  width = 4.3
)

      palette <- cmsCol
      # palette <- rev(c(brewer.pal(9, "Purples") [c( 6, 9)]))
      pal <- list(values=palette, na.value="gray80")
      
for(im in imSigs) {
  Others <- imSigs[imSigs != im]
  
  for (ot in Others) {
    
    d <- annMar %>%
      filter(! annMar$Sample_Type  %in%  "Normal") %>%
      mutate(medIm = median(!!sym(im)),
             medOt = median(!!sym(ot)))
    
    corVal <- round(cor.test(d[, im], d[, ot])[[4]], 2)
      
    print(
      d %>% 
        ggplot(.,
               # aes_string(im, ot, color = "MMR")) +
               aes_string(im, ot, color = "CMS")) +
        geom_point(alpha = 0.8) +
        invoke(scale_colour_manual, pal) +
        geom_vline(xintercept = d$medIm) +
        geom_hline(yintercept = d$medOt) +
        scale_x_continuous(
                       breaks = equal_breaks(n = nBreak, s = scalingFactor),
                       expand = c(scalingFactor, 0)
                     ) +
        scale_y_continuous(
                       breaks = equal_breaks(n = nBreak, s = scalingFactor),
                       expand = c(scalingFactor, 0)
                     ) +
        ggtitle(paste0("Marisa (cor = ", corVal, ")")) +
        currentTheme + 
        guides(color = guide_legend(nrow=2))
    )
  }
}

dev.off()

```





# OS

561 samples
```{r}
library(survival)
timeCol = "OS.time"
eventCol = "OS"
sn <- "OS"

annMarFilt <- annMarFilt[complete.cases(annMarFilt$Age), ]
annMarFilt <- annMarFilt[complete.cases(annMarFilt[, timeCol]), ]
annMarFilt$MMR[annMarFilt$MMR == "N/A"] <- NA
```
 

## Lasso for Cox to choose covariates
In order to decide which of the clinical factors to include, we need to perform lasso on the Cox models. We then use the coef of the lambda.min to decide which ones to include or exclude from the model. 
Based on the results below, we need to include all variables except for MMR.
```{r}
library(tidyverse)
library(broom)
library(glmnet)

set.seed(20201214)
annMarFilt_Sel <-
  annMarFilt[, c(
    "MMR",
    "CMS",
    "Age",
    "Stage",
    "OS.time",
    "OS"
  )]

## 442 patients after removing NAs
annMarFilt_Sel <- annMarFilt_Sel[complete.cases(annMarFilt_Sel) &
annMarFilt_Sel$OS.time > 0,]

sur <- Surv(
  time = as.numeric(annMarFilt_Sel[, timeCol]), 
  event = as.numeric(annMarFilt_Sel[, eventCol])) 


pred <- annMarFilt_Sel[, 
                          ! colnames(annMarFilt_Sel)  %in% c(timeCol, eventCol)]

x <- model.matrix(~ MMR + CMS + Age + Stage, pred)


y <- as.matrix(annMarFilt_Sel[, c(timeCol, eventCol)])
colnames(y) <- c("time", "status")

fit <- glmnet(
  x = x, 
  y = y, 
  family = "cox",
  alpha = 1)

plot(fit, label = T)
cv.fit <- cv.glmnet(x, y, family = "cox", alpha = 1)
plot(cv.fit)


coef(cv.fit, s = "lambda.min")


coxph(Surv(as.numeric(annMarFilt_Sel[, timeCol]), as.numeric(annMarFilt_Sel[, eventCol])) ~
        Age + Stage + MMR + CMS,
      data = annMarFilt_Sel)

coxph(Surv(as.numeric(annMarFilt_Sel[, timeCol]), as.numeric(annMarFilt_Sel[, eventCol])) ~
        Age + Stage + CMS,
      data = annMarFilt_Sel)

```

## Generate different format of the data
### Wide, 2/3 group scores/gene
```{r}

scoreColumns <- which(colnames(annMarFilt) %in% scoreNames)

wide2 <- data.frame(
  ## keep annotations and stage and age
  annMarFilt[, - c(scoreColumns)],
  ## add two group scores and genes
  as.data.frame(sapply(c(scoreNames), function(x) {
  case_when(
    annMarFilt[, x] >= median(annMarFilt[, x], na.rm = T) ~ "High",
    annMarFilt[, x] < median(annMarFilt[, x], na.rm = T) ~ "Low"
  )
})), check.names = F
)


wide3 <- data.frame(## keep annotations and stage and age
  annMarFilt[, - c(scoreColumns)],
  ## add two group scores and genes
  as.data.frame(sapply(c(scoreNames), function(x) {
    case_when(
      annMarFilt[, x] >= quantile(annMarFilt[, x], probs = 0.65) ~ "High",
      annMarFilt[, x] <= quantile(annMarFilt[, x], probs = 0.35) ~ "Low",
      
      annMarFilt[, x] > quantile(annMarFilt[, x], probs = 0.35) &
        annMarFilt[, x] < quantile(annMarFilt[, x], probs = 0.65)  ~ "Intermediate"
    )
  })) , check.names = F)
```

### Long 2/3 groups score/gene
```{r}
scoreAnnot <- pivot_longer(annMarFilt, 
                           cols = scoreColumns[1]:scoreColumns[length(scoreColumns)], 
                           values_to = "Score",
                           names_to = "Signature") %>% 
  data.frame()


scoreAnnot <- scoreAnnot %>%
  data.frame() %>%
  group_by(Signature) %>%
  # mutate(MedianScore = median(TotalScore)) %>%
  mutate(
    MedianScore = median(Score),
    Quantile_low = quantile(Score, probs = 0.35),
    Quantile_High = quantile(Score, probs = 0.65),
    Score_2status = case_when(
      Score >= MedianScore ~ "High score",
      Score < MedianScore ~ "Low score"
    ),
    Score_3status = case_when(
      Score >= Quantile_High ~ "High score",
      Score <= Quantile_low ~ "Low score",
      Score < Quantile_High &
        Score > Quantile_low ~ "Intermediate score"
    )
  ) %>%
  data.frame()

```


### OS: Score + covariates
Considering age and stage as covariate:
2 grouped scores: "CD8NK_Exh_Bulk" "NK_Exh_Bulk" 

3 groups: "DDR" "MYC_neg"  "NK_Exh_Bulk" "TP53_signaling"


```{r}
covarColumns <- c("Age", "Stage", "CMS")

cox_OS_Score_2groups <-
  coxph_multitest(
    data = scoreAnnot,
    variableNameCol = "Signature",
    variableCol = "Score_2status",
    timeCol = timeCol,
    eventCol = eventCol,
    nStrata = 2,
    returnCovarStat = F,
    covarCol = covarColumns,
    nCores = 10
  )


cox_OS_Score_2groups_sig <- cox_OS_Score_2groups %>% filter(`Pr(>|z|)` < 0.05) %>% 
  data.frame(check.names = F)

cox_OS_Score_2groups_sig$VariableName


write.csv(
  cox_OS_Score_2groups,
  paste0(outPath, "Marisa_Cox_OS_Score_2groups_AllStat.csv"),
  row.names = F
)


cox_OS_Score_3groups <-
  coxph_multitest(
    data = scoreAnnot,
    variableNameCol = "Signature",
    variableCol = "Score_3status",
    timeCol = timeCol,
    eventCol = eventCol,
    nStrata = 3,
    returnCovarStat = F,
    covarCol = covarColumns,
    nCores = 10
  )

cox_OS_Score_3groups_sig <- cox_OS_Score_3groups %>% filter(`Pr(>|z|)` < 0.05) %>% 
  data.frame(check.names = F)

cox_OS_Score_3groups_sig$VariableName

write.csv(
  cox_OS_Score_3groups,
  paste0(outPath, "Marisa_Cox_OS_Score_3groups_AllStat.csv"),
  row.names = F
)

```


#### Plot significant scores
```{r}

pdf(
  paste0(figPath, "OS_Marisa_Score2groups_Cox.pdf"),
  height = 4.5,
  width = 4.5
)

for (s in unique(cox_OS_Score_2groups_sig$VariableName)) {
  print(
    survival_plot(
      data =
        eMar[, rownames(annMarFilt)],
      # stratify = "expr",
      stratify = "score",
      annot = annMarFilt,
      scoreCol =  s,
      gene = NULL,
      covariate = NULL,
      isCategoricalCov = TRUE,
      timeCol = paste0(sn, ".time"),
      eventCol = sn,
      nGroup = 2,
      confInt = F,
      ylabel = paste0(sn),
      cols =
        rev(brewer.pal(10, "Paired"))
    )
  )
}

dev.off()

pdf(
  paste0(figPath, "OS_Marisa_Score3groups_Cox.pdf"),
  height = 4.5,
  width = 4.5
)

for (s in unique(cox_OS_Score_3groups_sig$VariableName)) {
  print(
    survival_plot(
      data =
        eMar[, rownames(annMarFilt)],
      # stratify = "expr",
      stratify = "score",
      annot = annMarFilt,
      scoreCol =  s,
      gene = NULL,
      covariate = NULL,
      isCategoricalCov = TRUE,
      timeCol = paste0(sn, ".time"),
      eventCol = sn,
      nGroup = 3,
      confInt = F,
      ylabel = paste0(sn),
      cols =
        rev(brewer.pal(10, "Paired"))
    )
  )
}

dev.off()
```

### OS: Score Pairs + covariate
```{r}
trmSigs <- c(
  "CD8_Res_Bulk",
  "CD4_Res_Bulk",
  "NK_Res_Bulk",
  "All_Res_Bulk", 
  "CD8NK_Res_Bulk"
  )

exhSigs <- c(
  "CD8_Exh_Bulk",
  "CD4_Exh_Bulk",
  "NK_Exh_Bulk",
  "All_Exh_Bulk",
  "CD8NK_Exh_Bulk"
  )


# selSigs <- c(trmSigs, exhSigs, TGFbGenes)

exhTGFbCan <- c(exhSigs, canSigs)
trmTGFbCan <- c(trmSigs, canSigs)

# chi_scorePairs_age_signifClean <- vector()
wide2Comb <- wide2

for(s in trmSigs){
  for(e in exhTGFbCan){
    currentColumn <- paste(s, e, sep = "--")
    wide2Comb[, currentColumn] <- paste(wide2Comb[, s], s , "&",
                                      wide2Comb[, e], e)
  }
} 
  
for(s in exhSigs){
  for(e in trmTGFbCan){
    currentColumn <- paste(s, e, sep = "--")
    wide2Comb[, currentColumn] <- paste(wide2Comb[, s], s , "&",
                                      wide2Comb[, e], e)
  }
}  


wide2Comb[, "CD8_Res_Bulk--CD4_Res_Bulk"] <-
  paste(wide2Comb[, "CD8_Res_Bulk"],
        "CD8_Res_Bulk" ,
        "&",
        wide2Comb[, "CD4_Res_Bulk"],
        "CD4_Res_Bulk")
 
wide2Comb[, "CD8_Res_Bulk--NK_Res_Bulk"] <-
  paste(wide2Comb[, "CD8_Res_Bulk"],
        "CD8_Res_Bulk" ,
        "&",
        wide2Comb[, "NK_Res_Bulk"],
        "NK_Res_Bulk")
 
wide2Comb[, "CD4_Res_Bulk--NK_Res_Bulk"] <-
  paste(wide2Comb[, "CD4_Res_Bulk"],
        "CD4_Res_Bulk" ,
        "&",
        wide2Comb[, "NK_Res_Bulk"],
        "NK_Res_Bulk")

 
wide2Comb[, "CD8_Exh_Bulk--CD4_Exh_Bulk"] <-
  paste(wide2Comb[, "CD8_Exh_Bulk"],
        "CD8_Exh_Bulk" ,
        "&",
        wide2Comb[, "CD4_Exh_Bulk"],
        "CD4_Exh_Bulk")
 
wide2Comb[, "CD8_Exh_Bulk--NK_Exh_Bulk"] <-
  paste(wide2Comb[, "CD8_Exh_Bulk"],
        "CD8_Exh_Bulk" ,
        "&",
        wide2Comb[, "NK_Exh_Bulk"],
        "NK_Exh_Bulk")
 
wide2Comb[, "CD4_Exh_Bulk--NK_Exh_Bulk"] <-
  paste(wide2Comb[, "CD4_Exh_Bulk"],
        "CD4_Exh_Bulk" ,
        "&",
        wide2Comb[, "NK_Exh_Bulk"],
        "NK_Exh_Bulk")


wide2CombLong <- wide2Comb %>% 
  pivot_longer(., cols = 156:ncol(wide2Comb), 
               values_to = "Comb_2status", names_to = "Comb_Name") %>% 
  data.frame(check.names = F)
  
```




```{r}
cox_OS_2Comb <-
  coxph_multitest(
    data = wide2CombLong,
    variableNameCol = "Comb_Name",
    variableCol = "Comb_2status",
    timeCol = timeCol,
    eventCol = eventCol,
    nStrata = 4,
    returnCovarStat = F,
    covarCol = covarColumns,
    nCores = 10
  )

cox_OS_2Comb_sig <- cox_OS_2Comb %>% filter(`Pr(>|z|)` < 0.05) %>% 
  data.frame(check.names = F)


# write.csv(
#   cox_OS_2Comb,
#   paste0(outPath, "Marisa_Cox_OS_2ScorePairs_AllStat2.csv"),
#   row.names = F
# )
# 
# write.csv(
#   cox_OS_2Comb_sig,
#   paste0(outPath, "Marisa_Cox_OS_2ScorePairs_significant2.csv"),
#   row.names = F
# )

```


```{r}
covarInfo <- paste(covarColumns, collapse = " + ")
sigPairs <- unique(cox_OS_2Comb_sig$VariableName)

pdf(
  paste0(figPath, "OS_Marisa_2ScorePairs_Cox_CorrectedCurves.pdf"),
  height = 4,
  width = 8
)

for(cc in sigPairs){
  
  variableCol <- "Comb_2status"
  ggSurv <- wide2CombLong[wide2CombLong$Comb_Name == cc,]
  currentCoxFit <-  coxph(as.formula(
    paste(
      'Surv(',
      timeCol,
      ', ',
      eventCol,
      ')~',
      variableCol,
      ' + ',
      covarInfo
    )
  )
  ,
  data = ggSurv)

  
    print(
    survminer::ggadjustedcurves(
      currentCoxFit,
      method = "average",
      variable = "Comb_2status",
      data = ggSurv, 
      palette = rev(brewer.pal(10, "Paired")), 
      title = cc, 
      legend = "right"
    ) 
  )
  
}

dev.off()  


pdf(
  paste0(figPath, "OS_Marisa_2ScorePairs_Cox_paper.pdf"),
  height = 5.2,
  width = 4.6
)
for(cc in sigPairs){
  
  sc1 <- unlist(strsplit(cc, "--"))[1]
  sc2 <- unlist(strsplit(cc, "--"))[2]
  
   print(survival_plot(data = 
                         eMar[, rownames(annMarFilt)],
                          # stratify = "score_score",
                          stratify = "score_score",
                          annot = annMarFilt,
                          # annot = currentAnnotExprCleanFilt,
                          # scoreCol = cc,
                          scoreCol = c(sc1, sc2),
                          gene = NULL,
                          covariate = NULL,
                          isCategoricalCov = NULL,
                          timeCol = paste0(sn, ".time"),
                          eventCol = sn,
                          nGroup = 2,
                          confInt = F, 
                          ylabel = sn,
                          cols = 
                            (brewer.pal(8, "Dark2"))))
}
dev.off()



# 
# pdf(
#   paste0(figPath, "OS_Marisa_2ScorePairs_Cox_AllScorePairs.pdf"),
#   height = 5.6,
#   width = 5.3
# )
# # for(cc in unique(cox_OS_2Comb_AgeStage$VariableName)){
# for(cc in unique(cox_OS_2Comb$VariableName)){
#   
#   sc1 <- unlist(strsplit(cc, "--"))[1]
#   sc2 <- unlist(strsplit(cc, "--"))[2]
#   
#    print(survival_plot(data = 
#                          eMar[, rownames(annMarFilt)],
#                           # stratify = "score_score",
#                           stratify = "score_score",
#                           annot = annMarFilt,
#                           # annot = currentAnnotExprCleanFilt,
#                           # scoreCol = cc,
#                           scoreCol = c(sc1, sc2),
#                           gene = NULL,
#                           covariate = NULL,
#                           isCategoricalCov = NULL,
#                           timeCol = paste0(sn, ".time"),
#                           eventCol = sn,
#                           nGroup = 2,
#                           confInt = F, 
#                           ylabel = sn,
#                           cols = 
#                             (brewer.pal(8, "Dark2"))))
# }
# dev.off()
```

### CoefPlot
```{r}
dd <- cox_OS_2Comb_sig[, c("VariableName",
                                    "variable",
                                    "exp(coef)",
                                    "lower .95",
                                    "upper .95",
                                    "z",
                                    "Pr(>|z|)")]

colnames(dd) <-
  c("VariableName", "term", "estimate", "conf.low", "conf.high", "z", "p-value")

# dd <- dd[! duplicated(dd$term),]
dd <- dd[order(dd$estimate), ]

  # geom_vline(xintercept = 1, lty = 2, size = 1, color = "darkblue") +
  # geom_vline(xintercept = 0, size = 1)


pdf(paste0(figPath, "OS_Marisa_ScorePairs_CoefPlot.pdf"),
    height = 14,
    width = 8)
ggstatsplot::ggcoefstats(
  x = dd,
  output = "plot",
  statistic = "z",
  exponentiate = T
)
dev.off()
```



### ScorePairs extremes
Get only top left and bottom right of teh scattreplot

```{r}

wide2CombLongMar <- wide2CombLong
selComb <- c(
  "Low CD8_Res_Bulk & High CD8_Exh_Bulk",
  "High CD8_Res_Bulk & Low CD8_Exh_Bulk",
  
  "Low CD4_Res_Bulk & High CD4_Exh_Bulk",
  "High CD4_Res_Bulk & Low CD4_Exh_Bulk",
  
  "Low NK_Res_Bulk & High NK_Exh_Bulk",
  "High NK_Res_Bulk & Low NK_Exh_Bulk",
  
  "Low All_Exh_Bulk & High All_Res_Bulk",
  "High All_Exh_Bulk & Low All_Res_Bulk",
  

  "High CD8_Res_Bulk & Low NK_Exh_Bulk",
  "Low CD8_Res_Bulk & High NK_Exh_Bulk",
  
  "High CD4_Res_Bulk & Low NK_Exh_Bulk",
  "Low CD4_Res_Bulk & High NK_Exh_Bulk",
  
  "High CD4_Res_Bulk & Low CD8_Exh_Bulk",
  "Low CD4_Res_Bulk & High CD8_Exh_Bulk",
  
  "High CD8_Exh_Bulk & Low NK_Res_Bulk",
  "Low CD8_Exh_Bulk & High NK_Res_Bulk",
  
  
  "High CD8_Exh_Bulk & Low NK_Exh_Bulk",
  "Low CD8_Exh_Bulk & High NK_Exh_Bulk",
  
  "High CD8_Exh_Bulk & Low CD4_Exh_Bulk",
  "Low CD8_Exh_Bulk & High CD4_Exh_Bulk",
  
  "High CD4_Exh_Bulk & Low NK_Exh_Bulk",
  "Low CD4_Exh_Bulk & High NK_Exh_Bulk",
  
  
  "High CD8_Res_Bulk & Low NK_Res_Bulk",
  "Low CD8_Res_Bulk & High NK_Res_Bulk",
  
  "High CD8_Res_Bulk & Low CD4_Res_Bulk",
  "Low CD8_Res_Bulk & High CD4_Res_Bulk",
  
  "High CD4_Res_Bulk & Low NK_Res_Bulk",
  "Low CD4_Res_Bulk & High NK_Res_Bulk", 
  

  "High CD8NK_Exh_Bulk & Low CD8NK_Res_Bulk",
  "Low CD8NK_Exh_Bulk & High CD8NK_Res_Bulk"
)

wide2CombLongMarSel <- wide2CombLongMar[wide2CombLongMar$Comb_2status  %in% selComb,]

cox_OS_2Comb_ExtremesMar <-
  coxph_multitest(
    data = wide2CombLongMarSel,
    variableNameCol = "Comb_Name",
    variableCol = "Comb_2status",
    timeCol = timeCol,
    eventCol = eventCol,
    nStrata = 2,
    returnCovarStat = F,
    covarCol = covarColumns,
    nCores = 10
  )

cox_OS_2Comb_Extremes_sigMar <- cox_OS_2Comb_ExtremesMar %>% filter(`Pr(>|z|)` <= 0.05) %>% 
  data.frame(check.names = F)


##----- 0.05:
# [1] "CD8_Trm_Bulk--CD8_Exh_Bulk" "CD4_Trm_Bulk--NK_Exh_Bulk" 
# [3] "CD8_Exh_Bulk--NK_Trm_Bulk"  "CD8_Exh_Bulk--CD4_Exh_Bulk"
# [5] "CD8_Trm_Bulk--NK_Exh_Bulk"  "NK_Trm_Bulk--NK_Exh_Bulk" 


pdf(
  paste0(figPath, "OS_Marisa_ExtremeScorePairs_Cox.pdf"),
  height = 5.8,
  width = 6
)

for (cc in cox_OS_2Comb_Extremes_sigMar$VariableName) {
  annot <-
    wide2CombLongMarSel[wide2CombLongMarSel$Comb_Name == cc, ]
  
  annot[, timeCol] <- as.numeric(annot[, timeCol])
  annot[, eventCol] <- as.numeric(annot[, eventCol])
  
  fitValues <- survfit(Surv(time = annot[, timeCol],
                            event = annot[, eventCol]) ~
                         annot$Comb_2status)
  
  nSamples <- as.numeric(table(annot$Comb_2status)[1])
  pValCox <- cox_OS_2Comb_Extremes_sigMar$`Pr(>|z|)`[cox_OS_2Comb_Extremes_sigMar$VariableName == cc]
  
  if (pValCox < 0.01) {
    pValCox <- " < 0.01"
  } else {
    pValCox <- paste0(" = ", round(pValCox, 3))
  }
  
  print(survminer::ggsurvplot (
    fitValues,
    data = annot,
    fun = "pct",
    pval = T,
    conf.int = F,
    palette = cols,
    xlab = "Time",
    legend.title = "",
    title = "Marisa",
    subtitle = paste0("p-value (Cox)", pValCox, ";", " N = ", nSamples, " in each group"),
    # legend.labs = c("High CD4_Exh & Low NK_Exh", "Low CD4_Exh & High NK_Exh"),
    legend = c(.5, .2),
    font.x = 20,
    font.y = 20,
    font.tickslab = 18,
    font.main = 20,
    font.submain = c(17, "italic"),
    font.legend = 11

  ))

}
dev.off()

```


## Save data list
```{r}

dListMar <- list(
  eMar = eMar,
  annMar = annMar,
  eMarFilt = eMarFilt,
  scoreAnnot = scoreAnnot,
  # exprAnnot = exprAnnot,
  annMarFilt = annMarFilt,
  wide2 = wide2,
  wide3 = wide3,
  wide2Comb = wide2Comb,
  wide2CombLong = wide2CombLong
  # wide2CombExpr = wide2CombExpr,
  # wide2CombExprLong = wide2CombExprLong
)


saveRDS(dListMar, paste0(outPath, "Marisa_SurvivalScoreExpr_DataFormatList_short_20201214.RDS"))

# dListMar <- readRDS( "../output/Microarray/Marisa_SurvivalScoreExpr_DataFormatList_short_20201214.RDS")

```
           
 



# RFS

559 samples
```{r}
library(survival)
timeCol = "RFS.time"
eventCol = "RFS"
sn <- "RFS"

annMarFilt <- annMarFilt[complete.cases(annMarFilt$Age), ]
annMarFilt <- annMarFilt[complete.cases(annMarFilt[, timeCol]), ]
annMarFilt$MMR[annMarFilt$MMR == "N/A"] <- NA
```
 
 
Remove samples that have time = 0; these are mostly stage IV samples that have recurrence (event = 1) and few stage II and III samples that have or do not have recurrence. In total 38 samples.
This time we keep the rest of the stage 4 samples in (30).
```{r}
# annMarFilt[(annMarFilt$RFS == 1 &
#                   annMarFilt$RFS.time == 0), 
#                c("Stage", "RFS.time", "RFS")]

## 521 samples
annMarFilt <- annMarFilt[
  ! annMarFilt$RFS.time == 0, 
]

```

## Lasso for Cox to choose covariates
In order to decide which of the clinical factors to include, we need to perform lasso on the Cox models. We then use the coef of the lambda.min to decide which ones to include or exclude from the model. 
Based on the results below, we need to include all 4 variables.
```{r}
library(tidyverse)
library(broom)
library(glmnet)

set.seed(20201214)
annMarFilt_Sel <-
  annMarFilt[, c(
    "MMR",
    "CMS",
    "Age",
    "Stage",
    "RFS.time",
    "RFS"
  )]

## 381 patients after removing NAs
annMarFilt_Sel <- annMarFilt_Sel[complete.cases(annMarFilt_Sel) &
annMarFilt_Sel$RFS.time > 0,]

sur <- Surv(
  time = as.numeric(annMarFilt_Sel[, timeCol]), 
  event = as.numeric(annMarFilt_Sel[, eventCol])) 


pred <- annMarFilt_Sel[, 
                          ! colnames(annMarFilt_Sel)  %in% c(timeCol, eventCol)]

x <- model.matrix(~ MMR + CMS + Age + Stage, pred)


y <- as.matrix(annMarFilt_Sel[, c(timeCol, eventCol)])
colnames(y) <- c("time", "status")

fit <- glmnet(
  x = x, 
  y = y, 
  family = "cox",
  alpha = 1)

plot(fit, label = T)
cv.fit <- cv.glmnet(x, y, family = "cox", alpha = 1)
plot(cv.fit)


coef(cv.fit, s = "lambda.min")


coxph(Surv(as.numeric(annMarFilt_Sel[, timeCol]), as.numeric(annMarFilt_Sel[, eventCol])) ~
        Age + Stage + MMR + CMS,
      data = annMarFilt_Sel)

coxph(Surv(as.numeric(annMarFilt_Sel[, timeCol]), as.numeric(annMarFilt_Sel[, eventCol])) ~
        Age + Stage + CMS,
      data = annMarFilt_Sel)

```

## Generate different format of the data
### Wide, 2/3 group scores/gene
```{r}
scoreColumns <- which(colnames(annMarFilt) %in% scoreNames)

wide2 <- data.frame(
  ## keep annotations and stage and age
  annMarFilt[, - c(scoreColumns)],
  ## add two group scores and genes
  as.data.frame(sapply(c(scoreNames), function(x) {
  case_when(
    annMarFilt[, x] >= median(annMarFilt[, x], na.rm = T) ~ "High",
    annMarFilt[, x] < median(annMarFilt[, x], na.rm = T) ~ "Low"
  )
})), check.names = F
)


wide3 <- data.frame(## keep annotations and stage and age
  annMarFilt[, - c(scoreColumns)],
  ## add two group scores and genes
  as.data.frame(sapply(c(scoreNames), function(x) {
    case_when(
      annMarFilt[, x] >= quantile(annMarFilt[, x], probs = 0.65) ~ "High",
      annMarFilt[, x] <= quantile(annMarFilt[, x], probs = 0.35) ~ "Low",
      
      annMarFilt[, x] > quantile(annMarFilt[, x], probs = 0.35) &
        annMarFilt[, x] < quantile(annMarFilt[, x], probs = 0.65)  ~ "Intermediate"
    )
  })) , check.names = F)
```

### Long 2/3 groups score/gene
```{r}
scoreAnnot <- pivot_longer(annMarFilt, 
                           cols = scoreColumns[1]:scoreColumns[length(scoreColumns)], 
                           values_to = "Score",
                           names_to = "Signature") %>% 
  data.frame()


scoreAnnot <- scoreAnnot %>%
  data.frame() %>%
  group_by(Signature) %>%
  # mutate(MedianScore = median(TotalScore)) %>%
  mutate(
    MedianScore = median(Score),
    Quantile_low = quantile(Score, probs = 0.35),
    Quantile_High = quantile(Score, probs = 0.65),
    Score_2status = case_when(
      Score >= MedianScore ~ "High score",
      Score < MedianScore ~ "Low score"
    ),
    Score_3status = case_when(
      Score >= Quantile_High ~ "High score",
      Score <= Quantile_low ~ "Low score",
      Score < Quantile_High &
        Score > Quantile_low ~ "Intermediate score"
    )
  ) %>%
  data.frame()


```



### RFS: Score + covariates
2 grouped scores: "Epi_Thiery"  "HPX_EGF_EMT" "Peroxisome"  "TP53_signaling"

3 groups: "HPX_EGF_EMT"  "Hypoxia"   "Hypoxia20"     
 "Peroxisome"  "TP53_signaling"


```{r}
covarColumns <- c("Age", "Stage", "CMS", "MMR")

cox_RFS_Score_2groups <-
  coxph_multitest(
    data = scoreAnnot,
    variableNameCol = "Signature",
    variableCol = "Score_2status",
    timeCol = timeCol,
    eventCol = eventCol,
    nStrata = 2,
    returnCovarStat = F,
    covarCol = covarColumns,
    nCores = 10
  )


cox_RFS_Score_2groups_sig <- cox_RFS_Score_2groups %>% filter(`Pr(>|z|)` < 0.05) %>% 
  data.frame(check.names = F)

cox_RFS_Score_2groups_sig$VariableName


write.csv(
  cox_RFS_Score_2groups,
  paste0(outPath, "Marisa_Cox_RFS_Score_2groups_AllStat.csv"),
  row.names = F
)


cox_RFS_Score_3groups <-
  coxph_multitest(
    data = scoreAnnot,
    variableNameCol = "Signature",
    variableCol = "Score_3status",
    timeCol = timeCol,
    eventCol = eventCol,
    nStrata = 3,
    returnCovarStat = F,
    covarCol = covarColumns,
    nCores = 10
  )

cox_RFS_Score_3groups_sig <- cox_RFS_Score_3groups %>% filter(`Pr(>|z|)` < 0.05) %>% 
  data.frame(check.names = F)

cox_RFS_Score_3groups_sig$VariableName

write.csv(
  cox_RFS_Score_3groups,
  paste0(outPath, "Marisa_Cox_RFS_Score_3groups_AllStat.csv"),
  row.names = F
)

```


#### Plot significant scores
```{r}

pdf(
  paste0(figPath, "RFS_Marisa_Score2groups_Cox.pdf"),
  height = 4.5,
  width = 4.5
)

for (s in unique(cox_RFS_Score_2groups_sig$VariableName)) {
  print(
    survival_plot(
      data =
        eMar[, rownames(annMarFilt)],
      # stratify = "expr",
      stratify = "score",
      annot = annMarFilt,
      scoreCol =  s,
      gene = NULL,
      covariate = NULL,
      isCategoricalCov = TRUE,
      timeCol = paste0(sn, ".time"),
      eventCol = sn,
      nGroup = 2,
      confInt = F,
      ylabel = paste0(sn),
      cols =
        rev(brewer.pal(10, "Paired"))
    )
  )
}

dev.off()

pdf(
  paste0(figPath, "RFS_Marisa_Score3groups_Cox.pdf"),
  height = 4.5,
  width = 4.5
)

for (s in unique(cox_RFS_Score_3groups_sig$VariableName)) {
  print(
    survival_plot(
      data =
        eMar[, rownames(annMarFilt)],
      # stratify = "expr",
      stratify = "score",
      annot = annMarFilt,
      scoreCol =  s,
      gene = NULL,
      covariate = NULL,
      isCategoricalCov = TRUE,
      timeCol = paste0(sn, ".time"),
      eventCol = sn,
      nGroup = 3,
      confInt = F,
      ylabel = paste0(sn),
      cols =
        rev(brewer.pal(10, "Paired"))
    )
  )
}

dev.off()
```



### RFS: Score Pairs + covariate
```{r}
trmSigs <- c(
  "CD8_Res_Bulk",
  "CD4_Res_Bulk",
  "NK_Res_Bulk",
  "All_Res_Bulk", 
  "CD8NK_Res_Bulk"
  )

exhSigs <- c(
  "CD8_Exh_Bulk",
  "CD4_Exh_Bulk",
  "NK_Exh_Bulk",
  "All_Exh_Bulk",
  "CD8NK_Exh_Bulk"
  )


exhTGFbCan <- c(exhSigs, canSigs)
trmTGFbCan <- c(trmSigs, canSigs)

wide2Comb <- wide2

for(s in trmSigs){
  for(e in exhTGFbCan){
    currentColumn <- paste(s, e, sep = "--")
    wide2Comb[, currentColumn] <- paste(wide2Comb[, s], s , "&",
                                      wide2Comb[, e], e)
  }
} 
  
for(s in exhSigs){
  for(e in trmTGFbCan){
    currentColumn <- paste(s, e, sep = "--")
    wide2Comb[, currentColumn] <- paste(wide2Comb[, s], s , "&",
                                      wide2Comb[, e], e)
  }
}  

wide2Comb[, "CD8_Res_Bulk--CD4_Res_Bulk"] <-
  paste(wide2Comb[, "CD8_Res_Bulk"],
        "CD8_Res_Bulk" ,
        "&",
        wide2Comb[, "CD4_Res_Bulk"],
        "CD4_Res_Bulk")
 
wide2Comb[, "CD8_Res_Bulk--NK_Res_Bulk"] <-
  paste(wide2Comb[, "CD8_Res_Bulk"],
        "CD8_Res_Bulk" ,
        "&",
        wide2Comb[, "NK_Res_Bulk"],
        "NK_Res_Bulk")
 
wide2Comb[, "CD4_Res_Bulk--NK_Res_Bulk"] <-
  paste(wide2Comb[, "CD4_Res_Bulk"],
        "CD4_Res_Bulk" ,
        "&",
        wide2Comb[, "NK_Res_Bulk"],
        "NK_Res_Bulk")

 
wide2Comb[, "CD8_Exh_Bulk--CD4_Exh_Bulk"] <-
  paste(wide2Comb[, "CD8_Exh_Bulk"],
        "CD8_Exh_Bulk" ,
        "&",
        wide2Comb[, "CD4_Exh_Bulk"],
        "CD4_Exh_Bulk")
 
wide2Comb[, "CD8_Exh_Bulk--NK_Exh_Bulk"] <-
  paste(wide2Comb[, "CD8_Exh_Bulk"],
        "CD8_Exh_Bulk" ,
        "&",
        wide2Comb[, "NK_Exh_Bulk"],
        "NK_Exh_Bulk")
 
wide2Comb[, "CD4_Exh_Bulk--NK_Exh_Bulk"] <-
  paste(wide2Comb[, "CD4_Exh_Bulk"],
        "CD4_Exh_Bulk" ,
        "&",
        wide2Comb[, "NK_Exh_Bulk"],
        "NK_Exh_Bulk")


wide2CombLong <- wide2Comb %>% 
  pivot_longer(., cols = 156:ncol(wide2Comb), 
               values_to = "Comb_2status", names_to = "Comb_Name") %>% 
  data.frame(check.names = F)
  
```




```{r}
cox_RFS_2Comb <-
  coxph_multitest(
    data = wide2CombLong,
    variableNameCol = "Comb_Name",
    variableCol = "Comb_2status",
    timeCol = timeCol,
    eventCol = eventCol,
    nStrata = 4,
    returnCovarStat = F,
    covarCol = covarColumns,
    nCores = 10
  )

cox_RFS_2Comb_sig <- cox_RFS_2Comb %>% filter(`Pr(>|z|)` < 0.05) %>% 
  data.frame(check.names = F)


# write.csv(
#   cox_RFS_2Comb,
#   paste0(outPath, "Marisa_Cox_RFS_2ScorePairs_AllStat.csv"),
#   row.names = F
# )
# 
# write.csv(
#   cox_RFS_2Comb_sig,
#   paste0(outPath, "Marisa_Cox_RFS_2ScorePairs_significant.csv"),
#   row.names = F
# )

```


```{r}

sigPairs <- unique(cox_RFS_2Comb_sig$VariableName)

pdf(
  paste0(figPath, "RFS_Marisa_2ScorePairs_Cox_CorrectedCurves.pdf"),
  height = 4,
  width = 8
)

for(cc in sigPairs){
  
  variableCol <- "Comb_2status"
  ggSurv <- wide2CombLong[wide2CombLong$Comb_Name == cc,]
  currentCoxFit <-  coxph(as.formula(
    paste(
      'Surv(',
      timeCol,
      ', ',
      eventCol,
      ')~',
      variableCol,
      ' + ',
      covarInfo
    )
  )
  ,
  data = ggSurv)

  
    print(
    survminer::ggadjustedcurves(
      currentCoxFit,
      method = "average",
      variable = "Comb_2status",
      data = ggSurv, 
      palette = rev(brewer.pal(10, "Paired")), 
      title = cc, 
      legend = "right"
    ) 
  )
  
}

dev.off()  


pdf(
  paste0(figPath, "RFS_Marisa_2ScorePairs_Cox_paper.pdf"),
  height = 5.2,
  width = 4.6
)
for(cc in sigPairs){
  
  sc1 <- unlist(strsplit(cc, "--"))[1]
  sc2 <- unlist(strsplit(cc, "--"))[2]
  
   print(survival_plot(data = 
                         eMar[, rownames(annMarFilt)],
                          # stratify = "score_score",
                          stratify = "score_score",
                          annot = annMarFilt,
                          # annot = currentAnnotExprCleanFilt,
                          # scoreCol = cc,
                          scoreCol = c(sc1, sc2),
                          gene = NULL,
                          covariate = NULL,
                          isCategoricalCov = NULL,
                          timeCol = paste0(sn, ".time"),
                          eventCol = sn,
                          nGroup = 2,
                          confInt = F, 
                          ylabel = sn,
                          cols = 
                            (brewer.pal(8, "Dark2"))))
}
dev.off()


# 
# pdf(
#   paste0(figPath, "RFS_Marisa_2ScorePairs_Cox_AllScorePairs.pdf"),
#   height = 5.6,
#   width = 5.3
# )
# for(cc in unique(cox_RFS_2Comb$VariableName)){
# 
#   sc1 <- unlist(strsplit(cc, "--"))[1]
#   sc2 <- unlist(strsplit(cc, "--"))[2]
# 
#    print(survival_plot(data =
#                          eMar[, rownames(annMarFilt)],
#                           # stratify = "score_score",
#                           stratify = "score_score",
#                           annot = annMarFilt,
#                           # annot = currentAnnotExprCleanFilt,
#                           # scoreCol = cc,
#                           scoreCol = c(sc1, sc2),
#                           gene = NULL,
#                           covariate = NULL,
#                           isCategoricalCov = NULL,
#                           timeCol = paste0(sn, ".time"),
#                           eventCol = sn,
#                           nGroup = 2,
#                           confInt = F,
#                           ylabel = sn,
#                           cols =
#                             (brewer.pal(8, "Dark2"))))
# }
# dev.off()
```

### CoefPlot
```{r}
dd <- cox_RFS_2Comb_sig[, c("VariableName",
                                    "variable",
                                    "exp(coef)",
                                    "lower .95",
                                    "upper .95",
                                    "z",
                                    "Pr(>|z|)")]

colnames(dd) <-
  c("VariableName", "term", "estimate", "conf.low", "conf.high", "z", "p-value")

dd <- dd[order(dd$estimate), ]

pdf(paste0(figPath, "RFS_Marisa_ScorePairs_CoefPlot.pdf"),
    height = 10,
    width = 6.5)
ggstatsplot::ggcoefstats(
  x = dd,
  output = "plot",
  statistic = "z",
  exponentiate = T
)
dev.off()
```



### ScorePairs extremes
Get only top left and bottom right of the scattreplot

```{r}
wide2CombLongMar <- wide2CombLong
selComb <- c(
  "Low CD8_Res_Bulk & High CD8_Exh_Bulk",
  "High CD8_Res_Bulk & Low CD8_Exh_Bulk",
  
  "Low CD4_Res_Bulk & High CD4_Exh_Bulk",
  "High CD4_Res_Bulk & Low CD4_Exh_Bulk",
  
  "Low NK_Res_Bulk & High NK_Exh_Bulk",
  "High NK_Res_Bulk & Low NK_Exh_Bulk",
  
  "Low All_Exh_Bulk & High All_Res_Bulk",
  "High All_Exh_Bulk & Low All_Res_Bulk",


  "High CD8_Res_Bulk & Low NK_Exh_Bulk",
  "Low CD8_Res_Bulk & High NK_Exh_Bulk",
  
  "High CD4_Res_Bulk & Low NK_Exh_Bulk",
  "Low CD4_Res_Bulk & High NK_Exh_Bulk",
  
  "High CD4_Res_Bulk & Low CD8_Exh_Bulk",
  "Low CD4_Res_Bulk & High CD8_Exh_Bulk",
  
  "High CD8_Exh_Bulk & Low NK_Res_Bulk",
  "Low CD8_Exh_Bulk & High NK_Res_Bulk",
  
  
  "High CD8_Exh_Bulk & Low NK_Exh_Bulk",
  "Low CD8_Exh_Bulk & High NK_Exh_Bulk",
  
  "High CD8_Exh_Bulk & Low CD4_Exh_Bulk",
  "Low CD8_Exh_Bulk & High CD4_Exh_Bulk",
  
  "High CD4_Exh_Bulk & Low NK_Exh_Bulk",
  "Low CD4_Exh_Bulk & High NK_Exh_Bulk",
  
  
  "High CD8_Res_Bulk & Low NK_Res_Bulk",
  "Low CD8_Res_Bulk & High NK_Res_Bulk",
  
  "High CD8_Res_Bulk & Low CD4_Res_Bulk",
  "Low CD8_Res_Bulk & High CD4_Res_Bulk",
  
  "High CD4_Res_Bulk & Low NK_Res_Bulk",
  "Low CD4_Res_Bulk & High NK_Res_Bulk", 
  

  "High CD8NK_Exh_Bulk & Low CD8NK_Res_Bulk",
  "Low CD8NK_Exh_Bulk & High CD8NK_Res_Bulk"
)

wide2CombLongMarSel <- wide2CombLongMar[wide2CombLongMar$Comb_2status  %in% selComb,]

cox_RFS_2Comb_ExtremesMar <-
  coxph_multitest(
    data = wide2CombLongMarSel,
    variableNameCol = "Comb_Name",
    variableCol = "Comb_2status",
    timeCol = timeCol,
    eventCol = eventCol,
    nStrata = 2,
    returnCovarStat = F,
    covarCol = covarColumns,
    nCores = 10
  )

cox_RFS_2Comb_Extremes_sigMar <- cox_RFS_2Comb_ExtremesMar %>% filter(`Pr(>|z|)` <= 0.05) %>% 
  data.frame(check.names = F)


##----- 0.05:
# "CD8_Exh_Bulk--NK_Trm_Bulk"
#  "NK_Trm_Bulk--NK_Exh_Bulk" 

pdf(
  paste0(figPath, "RFS_Marisa_ExtremeScorePairs_Cox.pdf"),
  height = 5.8,
  width = 6
)

for (cc in cox_RFS_2Comb_Extremes_sigMar$VariableName) {
  annot <-
    wide2CombLongMarSel[wide2CombLongMarSel$Comb_Name == cc, ]
  
  annot[, timeCol] <- as.numeric(annot[, timeCol])
  annot[, eventCol] <- as.numeric(annot[, eventCol])
  
  fitValues <- survfit(Surv(time = annot[, timeCol],
                            event = annot[, eventCol]) ~
                         annot$Comb_2status)
  
  nSamples <- as.numeric(table(annot$Comb_2status)[1])
  pValCox <- cox_RFS_2Comb_Extremes_sigMar$`Pr(>|z|)`[
    cox_RFS_2Comb_Extremes_sigMar$VariableName == cc]
  
  if (pValCox < 0.01) {
    pValCox <- " < 0.01"
  } else {
    pValCox <- paste0(" = ", round(pValCox, 3))
  }
  
  print(survminer::ggsurvplot (
    fitValues,
    data = annot,
    fun = "pct",
    pval = T,
    conf.int = F,
    palette = cols,
    xlab = "Time",
    legend.title = "",
    title = "Marisa (RFS)",
    subtitle = paste0("p-value (Cox)", pValCox, ";", " N = ", nSamples, " in each group"),
    # legend.labs = c("High CD4_Exh & Low NK_Exh", "Low CD4_Exh & High NK_Exh"),
    legend = c(.5, .2),
    font.x = 20,
    font.y = 20,
    font.tickslab = 18,
    font.main = 20,
    font.submain = c(17, "italic"),
    font.legend = 11

  ))

}
dev.off()

```


## Save data list
```{r}

dListMar_RFS <- list(
  eMar = eMar,
  annMar = annMar,
  eMarFilt = eMarFilt,
  scoreAnnot = scoreAnnot,
  # exprAnnot = exprAnnot,
  annMarFilt = annMarFilt,
  wide2 = wide2,
  wide3 = wide3,
  wide2Comb = wide2Comb,
  wide2CombLong = wide2CombLong
  # wide2CombExpr = wide2CombExpr,
  # wide2CombExprLong = wide2CombExprLong
)


saveRDS(dListMar_RFS, paste0(outPath, "Marisa_SurvivalRFS_ScoreExpr_DataFormatList_short_20201214.RDS"))

# dListMar <- readRDS( "../output/Microarray/Marisa_SurvivalRFS_ScoreExpr_DataFormatList_short_20201214.RDS")

```
          
          
          
        
# Mutation
```{r}

##---- if plotting ExhTrmScores
# annMarLong2 <- annMarLong %>%
#   pivot_longer(cols = c(BRAF, KRAS, TP53), values_to = "Mut", names_to = "Genes") %>%
#   filter(Signature %in% c(imSigs, canSigs))

annMarLong2 <- annMarLong %>%
  pivot_longer(cols = c(BRAF, TP53, KRAS), values_to = "Mut", names_to = "Genes") %>%
  # filter(Signature %in% c("All_Exh", "All_Trm"))
  filter(grepl("Bulk", Signature))


##---- if plotting Cancer scores
# annMarLong2 <- annMarLong %>% 
#   pivot_longer(cols = c(BRAF, KRAS, TP53), values_to = "Mut", names_to = "Genes") %>% 
#   filter(Signature %in% canSigs)

##--------

annMarLong2$Gene_Signature <- paste(annMarLong2$Genes, annMarLong2$Signature, sep = "_")

annMarLong2$Mut[annMarLong2$Mut == "M"] <- "Mut"
annMarLong2$Mut <- factor(annMarLong2$Mut, levels = c("WT", "Mut"))

currentcombs <- unique(annMarLong2$Gene_Signature)[order(unique(annMarLong2$Gene_Signature))]

p <- list()

for(i in 1:length(currentcombs)) {
  
  g <- unique(annMarLong2$Gene_Signature)[i]
  
  currentAnnMarLong2 <- annMarLong2 %>% 
    filter(Gene_Signature == g)

  ylab <- unique(currentAnnMarLong2$Signature)
  currentGene <- as.character(unique(currentAnnMarLong2$Genes))
  title <- currentGene
    # paste0(currentGene,"\n(p < 0.01)")
  
  p[[g]] <- currentAnnMarLong2 %>%
    filter(Sample_Type == "Cancer" &
             Mut != "N/A") %>%
    ggplot(data = ., aes(x = Mut, y = Score, color = Mut)) +
    geom_boxplot() +
    geom_jitter(position = position_jitter(0.2), alpha = 0.4) +
    ylab(ylab) +
  geom_signif(comparisons = list(c("Mut", "WT")),
              map_signif_level = TRUE, tip_length = 0, vjust= 1.4) +
    # geom_line(mapping = aes(y = .fitted), color = "blue") +
    ggtitle(title) +
    # scale_color_manual(values=c("#999999", "#E69F00")) +
    scale_color_manual(values = brewer.pal(4, "Dark2")[c(1, 3)]) +
    # scale_x_discrete(breaks = c("WT", "Mut"),
    #                  labels = c("WT", "Mut")) +
    scale_y_continuous(
      breaks = equal_breaks(n = nBreak, s = scalingFactor),
      expand = c(scalingFactor, 0)
    ) +
    theme_bw() +
    theme(axis.title.x = element_blank(), legend.position = "none")
}



library(gridExtra)
p <- p[currentcombs]
# pdf(paste0(figPath, "Boxplot_Marisa_Muts_CancerScores_sigs2.pdf"), height = 15, width = 13)
pdf(paste0(figPath, "Boxplot_Marisa_Muts_ScoresAllExhTrm_3Genes.pdf"), height = 7.5, width = 13)
# do.call(grid.arrange, p)
grid.arrange(
  grobs = p,
  ncol = 10
  # widths = c(1, 1, 1),
  # layout_matrix = rbind(c(1, 2, 3))
)
dev.off()

```

  
          
          
          
          
          



















