---
title: 'Test the NK cell Res and Exh signatures in several single cell data sets (CRC and normal gut)'
author: 
  - name: 'Momeneh (Sepideh) Foroutan'
    affiliation: 'Huntington Cancer Immunotherapy Lab'
    url: https://www.monash.edu/discovery-institute/huntington-lab
date: '12-06-2020'
output:
  rmdformats::readthedown:
    fig_width: 12
    fig_height: 6
    gallery: TRUE
    highlight: tango
    lightbox: TRUE
    self_contained: TRUE
    thumbnails: FALSE
    number_sections: TRUE	
    toc_depth: 4
    use_bookdown: TRUE
    code_folding: hide
  html_document2:
    df_print: paged
params:
  update_date: !r paste("Last updated on:", Sys.Date() )
editor_options: 
  chunk_output_type: inline
---
`r params$update_date`

<style>
body {
text-align: justify}
</style>


```{r}
library(knitr)
library(rmdformats)

options(max.print = "75")

opts_chunk$set(
  # echo = FALSE,
  cache = TRUE,
  # prompt = FALSE,
  # tidy = FALSE,
  comment = NA,
  message = FALSE,
  warning = FALSE,
  # results = FALSE,
  root.dir = normalizePath("."))

opts_knit$set(width = 65)
```

# Set up and overview
This document contains the codes used to generate some of the results in the [Foroutan et al](https://www.biorxiv.org/content/10.1101/2020.12.19.423379v1) manuscript. In this document, we test the Res and Exh signatures that we obtained by analysing the NK single cell data from [Zhang et al (Smart-seq2)](https://pubmed.ncbi.nlm.nih.gov/32302573/) in multiple independent single cell data sets. These include the CRC infiltrated NK cells from [Zhang et al 10X data](https://pubmed.ncbi.nlm.nih.gov/32302573/) and [de Vries et al data](https://pubmed.ncbi.nlm.nih.gov/31270164/), as well as the normal gut infiltrated NK cells from [James et al data](https://www.nature.com/articles/s41590-020-0602-z).

```{r, warning=F, message=F}

library(DT)
library(Seurat)
library(SingleCellExperiment)
library(singscore)
library(slingshot)
library(tidyverse)
library(RColorBrewer)

mainDir <- getwd()
outPath <- "../output/NK/"
figPath <- "../figure/NK/"
dataPath <- "../data/"
scriptPath <- "../script/"

tpmPath <- "../data/GSE146771_Zhang_CRC/"
devriesPath <- "../data/deVries/"
JamesPath <- "../data/James_GutAtlas/"

# ifelse(!dir.exists(file.path(mainDir, outPath)), dir.create(file.path(mainDir, outPath)), FALSE)
# ifelse(!dir.exists(file.path(mainDir, figPath)), dir.create(file.path(mainDir, figPath)), FALSE)
# ifelse(!dir.exists(file.path(mainDir, dataPath)), dir.create(file.path(mainDir, dataPath)), FALSE)
# ifelse(!dir.exists(file.path(mainDir, scriptPath)), dir.create(file.path(mainDir, scriptPath)), FALSE)


options(digits = 3)

equal_breaks <- function(n = nBreak, s = scalingFactor, ...){
  function(x){
    # rescaling
    d <- s * diff(range(x)) / (1+2*s)
    round( seq(min(x)+d, max(x)-d, length=n), 2)
  }
}

nBreak = 3
scalingFactor = 0.05

textSize <- 1.2

currentTheme <- theme_bw() +
  theme(
    # panel.background = element_blank()
    axis.title = element_text(size = rel(textSize)),
    axis.text = element_text(angle = 0, size = rel(textSize)),
    # strip.background = element_rect(colour = "#f0f0f0", fill = "#f0f0f0"),
    strip.text = element_text(size = rel(textSize)),
    axis.line = element_line(colour = "black", size = 0.5),
    legend.position = 'top',
    legend.title = element_text(size = rel(textSize), face = "italic"),
    legend.text = element_text(size = rel(textSize)),
    legend.key.size = unit(1, 'lines'),
    ## increase the line space
    plot.title = element_text(
      face = "bold",
      size = rel(textSize),
      hjust = 0.5
    )
  )

cols <-  c(
  brewer.pal(8, "Dark2")[-5],
  brewer.pal(10, "Paired"),
  brewer.pal(12, "Set3"),
  brewer.pal(9, "Blues")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Oranges")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Greens")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Purples")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Reds")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "Greys")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "BuGn")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "PuRd")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "BuPu")[c(8, 3, 7, 4, 6, 9, 5)],
  brewer.pal(9, "YlGn")[c(8, 3, 7, 4, 6, 9, 5)])


source(paste0(scriptPath, "visReduction_function.R"))



metaSmart <-
  read.table(
    paste0(tpmPath, "GSE146771_CRC.Leukocyte.Smart-seq2.Metadata.txt"),
    header = T,
    sep = "\t"
  )

metaSmart$Sub_Cluster <- as.character(metaSmart$Sub_Cluster)
metaSmart$Sub_Cluster <- sapply(metaSmart$Sub_Cluster, function(x){
  unlist(strsplit(x, "_"))[2]
})


immCells <- metaSmart$Sub_Cluster[grepl("CD8", metaSmart$Sub_Cluster)|
                                    grepl("NK", metaSmart$Sub_Cluster) |
                                    grepl("CD4", metaSmart$Sub_Cluster)|
                                    grepl("ILC3", metaSmart$Sub_Cluster)]

cellCols <- cols[1:length(unique(immCells))]
names(cellCols) <- unique(immCells)[order(unique(immCells))]

nkCols <- cellCols[grepl("NK-", names(cellCols))]
```


## Read signatures
Read the Res and Exh signatures that we obtained in the Smart-seq2 NK cell.

```{r}
nkSigs <-
  read.table(
    paste0(outPath, "ResExhMarkers_NK_CLs_LCM.txt"),
    header = T,
    sep = "\t",
    stringsAsFactors = F,
    check.names = F
  )

NK_ExhMarker <- nkSigs$gene[nkSigs$Marker == "Exh"]
NK_ResMarker <- nkSigs$gene[nkSigs$Marker == "Res"]

```

# Zhang 10X data
We ran the below code, which is commented now,to read the original 10X data (GSE146771_CRC.Leukocyte.10x.TPM.txt) and generate the Seurat object RData file for the NK cells in that data. Here we directly read the Seurat object.
```{r}
# crc10x <-
#   data.table::fread(
#     paste0(tpmPath, "GSE146771_CRC.Leukocyte.10x.TPM.txt")
#   )
# 
# rnames <- crc10x$V1
# crc10x <- crc10x[, 2:ncol(crc10x)]
# rownames(crc10x) <- rnames
# 
# crc10x <- as.matrix(crc10x)
# rownames(crc10x) <- rnames
# 
# 
# meta10x <-
#   read.table(
#     paste0(tpmPath, "GSE146771_CRC.Leukocyte.10x.Metadata.txt"),
#     header = T,
#     sep = "\t"
#   )
# 
# 
# meta10x$Sub_Cluster <- as.character(meta10x$Sub_Cluster)
# meta10x$Sub_Cluster <- sapply(meta10x$Sub_Cluster, function(x){
#   unlist(strsplit(x, "_"))[2]
# })
# 
# 
# sub10xNK <- crc10x[, grepl("NK-", meta10x$Sub_Cluster) ]
# 
# subMeta10xNK <- meta10x[grepl("NK-", meta10x$Sub_Cluster) , ]
# 
# rownames(subMeta10xNK) <- subMeta10xNK$CellName
# 
# all(subMeta10xNK$CellName == colnames(sub10xNK))
# 
# nk_10x <-
#   CreateSeuratObject(counts = sub10xNK,
#                      project = "Zhang_10X_NK",
#                      meta.data = subMeta10xNK,
#                      min.cells = 5)
# 
# nk_10x[["percent.mt"]] <- PercentageFeatureSet(nk_10x, pattern = "^MT-")
# 
# cell_attrsNK_10x <- nk_10x[[]]

# saveRDS(nk_10x, file = paste0(outPath, "DontPush/Zhang_10X_SeuratObj_NK_20200901.RDS"))


nk_10x <- readRDS(paste0(outPath, "DontPush/Zhang_10X_SeuratObj_NK_20200901.RDS"))

nk_10x$MSI <- "MSS"
nk_10x$MSI[nk_10x$Sample %in% c("P0123", "P0413", "P0825")] <- "MSI-H"
```



```{r, fig.height = 4, fig.width = 9, fig.cap = "Relationship between percent mito genes or number of detected genes and library size (from Seurat)"}

plot1 <-
  FeatureScatter(nk_10x,
                 feature1 = "nCount_RNA",
                 feature2 = "percent.mt",
                 group.by = "Sub_Cluster", 
                 cols = nkCols)
plot2 <-
  FeatureScatter(nk_10x,
                 feature1 = "nCount_RNA",
                 feature2 = "nFeature_RNA",
                 group.by = "Sample", 
                 cols = cols)
plot1 + plot2

```


```{r, fig.height=8, fig.width = 4, fig.cap = "Number of features, library size and percent mito genes in NK cells across patients"}
VlnPlot(
  nk_10x,
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
  ncol = 1,
  group.by = "Sample",
  pt.size = 0, cols = cols
)
```

### Standard Seurat pipeline
```{r, cache = T, warning = F, message = F}
nk_10x <- FindVariableFeatures(nk_10x, selection.method = "vst", nfeatures = 2000)

all.genes <- rownames(nk_10x)
nk_10x <- ScaleData(nk_10x, features = all.genes)

nk_10x <-
  RunPCA(nk_10x,
         features = VariableFeatures(object = nk_10x),
         verbose = F)

# DimHeatmap(nk_10x, dims = 1:30, cells = 500, balanced = TRUE)
# ElbowPlot(nk_10x, ndims = 40)

nk_10x <- FindNeighbors(nk_10x, dims = 1:10)
nk_10x <- FindClusters(nk_10x, resolution = 0.5, random.seed = 20200901)

nk_10x <- RunUMAP(nk_10x, dims = 1:10, n.neighbors = 30, min.dist = 0.3, seed.use = 20200901)
```

### Highly Variable genes
```{r, fig.cap = "Highly variable genes"}
top10 <- head(VariableFeatures(nk_10x), 20)

plot1 <- VariableFeaturePlot(nk_10x)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```

```{r, fig.height = 6, fig.width = 6, fig.cap = "Gene loadings for PC1 and PC2"}
VizDimLoadings(nk_10x, dims = 1:2, reduction = "pca")
```


### Score 10X NK cells
```{r, message = F, warning = F}
rd_10X <- rankGenes(as.matrix(GetAssayData(nk_10x)))

nk_10x$NK_Res <- simpleScore(
  rankData = rd_10X,
  upSet =  NK_ResMarker,
  knownDirection = T,
  centerScore = F
)$TotalScore

nk_10x$NK_Exh <- simpleScore(
  rankData = rd_10X,
  upSet =  NK_ExhMarker,
  knownDirection = T,
  centerScore = F
)$TotalScore
```


## UMAP/PCA based on markers
```{r, cache = T, warning=F, message=F}
gg <-  c(NK_ExhMarker, NK_ResMarker)

NKExhTrm_10x <- RunPCA(nk_10x, features = gg, verbose = F)

NKExhTrm_10x <- FindNeighbors(NKExhTrm_10x, reduction = "pca", dims = 1:5)
NKExhTrm_10x <- FindClusters(NKExhTrm_10x, resolution = 0.3)

NKExhTrm_10x <- RunUMAP(NKExhTrm_10x, reduction = "pca", 
          dims = 1:5,  min.dist = 0.2, n.neighbors = 40)

# DimPlot(NKExhTrm_10x, reduction = "umap")
# DimPlot(NKExhTrm_10x, reduction = "umap", group.by = "Sub_Cluster")
# DimPlot(NKExhTrm_10x, reduction = "umap", group.by = "Tissue")
# FeaturePlot(NKExhTrm_10x, reduction = "umap", features = "NK_Exh")
# FeaturePlot(NKExhTrm_10x, reduction = "umap", features = "NK_Res")
```


## UMAP plots coloured by annotations and scores
We colour the UMAPs generated using our Res and Exh marker genes based on different annotations, including the cell annotations (reported by Zhang et al), Seurat clusters (obtained using our marker genes), cell sources (P: peripheral blood, T: tumour, and N: normal), and samples (patients). Then, we colour the same UMAPs using the Res and Exh scores.
```{r, fig.height=4.5, fig.width=4.2, fig.cap="UMAP plots genearted by Res and Exh genes, and coloured by annotations and scores"}
anns <- c("seurat_clusters", "Tissue", "Sample")

# pdf(
#   paste0(figPath, "UMAP_Zhang10X_NK_ResExhMarkers_20201005_CellAnnot.pdf"),
#   height = 4,
#   width = 3.8
# )

DimPlot(NKExhTrm_10x, 
                reduction = "umap", 
        group.by = "Sub_Cluster") +
  theme_minimal() +
  scale_color_manual(values = nkCols) +
  theme(legend.position = "top") +
  guides(color = guide_legend(
    nrow=3, 
    override.aes = list(size=3)))

for(a in anns){
  print(DimPlot(NKExhTrm_10x, 
                reduction = "umap", 
        group.by = a) +
  theme_minimal() +
  scale_color_manual(values = cols) +
  theme(legend.position = "top") +
  guides(color = guide_legend(
    nrow=3, 
    override.aes = list(size=3))))
}

FeaturePlot(NKExhTrm_10x,
            "NK_Exh",
            pt.size = 0.7,
            order = T
            ) +
  scale_color_viridis_c(100) +
  theme_dark() +
  theme(legend.position = "top")

FeaturePlot(NKExhTrm_10x,
            "NK_Res",
            pt.size = 0.7,
            order = T
            ) +
  scale_color_viridis_c(100) +
  theme_dark() +
  theme(legend.position = "top")

# dev.off()

```

Blend plot coloured by scores (related to Figure 2 in the paper)
```{r, fig.height=3, fig.width=12}
# pdf(
#   paste0(figPath, "UMAP_Zhang10X_Blend_ExhResMarkers_NK_20201005.pdf"),
#   height = 3,
#   width = 12
# )
FeaturePlot(NKExhTrm_10x,
            features = c("NK_Res", "NK_Exh"),
            pt.size = 0.7,
            order = T, 
            blend = T,
            blend.threshold = 0.5,
            cols =  c("grey90", "orange2", "blue"))

# dev.off()
```

```{r}
# Save NK data after UMAP/Scores
# saveRDS(nk_10x, file = paste0(outPath, "DontPush/Zhang_10X_SeuratObj_NK_20201005_UMAP_Scores.RDS"))
```


# de Vries data
The raw data (matrix, genes, barcodes) and cell annotations were received from the authors of the [de Vries et al paper](https://pubmed.ncbi.nlm.nih.gov/31270164/), and then we generated a Seurat object. Here we read the Seurat object (33694 features across 1079 cells). 
```{r}
# vries <- Read10X(data.dir = devriesPath)
# vv0 <- CreateSeuratObject(vries)
# 
# vAnn <- read.csv(paste0(devriesPath, "celltoclusterID.CSV"),
#                  stringsAsFactors = F)
# colnames(vAnn) <- c("ID", "CellType")
# rownames(vAnn) <- paste(vAnn$ID, "1", sep = "-")
# 
# vv0 <- vv0[, rownames(vAnn)  ]
# 
# all(rownames(vAnn) == colnames(vv0))
# 
# vv0@meta.data <- vAnn
# saveRDS(vv0, paste0(devriesPath, "deVries_CRC_SeuratObject.RDS"))

vv0 <- readRDS(paste0(devriesPath, "deVries_CRC_SeuratObject.RDS"))
```


## Subset to NK/ILC and Others and Prolif:
```{r}
getCells <-
  c(
    "ILC/NK cells",
    "Proliferating cells",
    "Others"
  )

v <- vv0[, vv0$CellType %in% getCells]
# grep("^MT-", rownames(v), value = T)
v[["percent.mt"]] <- PercentageFeatureSet(v, pattern = "^MT-")

```


## Counts vs percent mt or number of features
```{r, fig.height = 4, fig.width = 9, fig.cap = "Relationship between percent mito genes or number of detected genes and library size (from Seurat)"}

plot1 <- FeatureScatter(v, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "CellType")
plot2 <- FeatureScatter(v, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "CellType")
plot1 + plot2

```

### Filter cells 
We only keep cells that have count of less than 15,000 (184 cells).
```{r}
## remove 1 cells --> 184 cells
v <- v[, v$nCount_RNA < 15000]
```


```{r, fig.height=8, fig.width = 4, fig.cap = "Number of features, library size and percent mito genes in NK cells across patients"}
VlnPlot(v, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 1, group.by = "CellType", pt.size = 0)
```



## Normalise using SCTransform
```{r, warning = F, message = F}
# library(sctransform)
v <- SCTransform(
  object = v,
  assay = "RNA",
  new.assay.name = "SCT",
  do.correct.umi = TRUE,
  ncells = NULL,
  variable.features.n = 5000,
  # variable.features.rv.th = 1.3,
  vars.to.regress = "percent.mt",  ## NULL by default
  do.scale = FALSE,
  do.center = TRUE,
  conserve.memory = FALSE,
  return.only.var.genes = TRUE,
  seed.use = 20201005,
  verbose = FALSE
)
```



```{r, fig.width = 9, fig.height = 4, fig.cap = "Mean-variance relationship after normalisation, annotated for highly variable genes"}
gg <- head(VariableFeatures(v), 15)
plot1 <- VariableFeaturePlot(v)
plot2 <- LabelPoints(plot = plot1, points = gg, repel = TRUE)
plot2
# or:
# CombinePlots(list(plot1,plot2))
```

For visualisation of the UMAp, etc we could use the SCT normalised data.
```{r}
v <- RunPCA(v, verbose = FALSE)

v <- FindNeighbors(v, dims = 1:15)
v <- FindClusters(v, resolution = 0.5, random.seed = 20201005)

v <- RunUMAP(v, dims = 1:15, n.neighbors = 20, min.dist = 0.2, seed.use = 20201005)

# DimPlot(v, reduction = "umap",
#         group.by = "CellType") +
#   theme_minimal() +
#   scale_color_manual(values = cols) +
#   theme(legend.position = "top") +
#   guides(color = guide_legend(
#     nrow=1, 
#     override.aes = list(size=3)))
```


## RNA slot
```{r}
vRNA <- v

DefaultAssay(vRNA) <- "RNA"
# Normalize RNA data for visualization purposes
vRNA <- NormalizeData(vRNA, verbose = FALSE)
vRNA <- FindVariableFeatures(vRNA, selection.method = "vst", nfeatures = 1000)
all.genes <- rownames(vRNA)
vRNA <- ScaleData(vRNA, features = all.genes)

vRNA <- RunPCA(vRNA, features = VariableFeatures(object = vRNA), verbose = F)
ElbowPlot(vRNA, ndims = 40)
vRNA <- FindNeighbors(vRNA, dims = 1:20)
vRNA <- FindClusters(vRNA, resolution = 0.5)
vRNA <- RunUMAP(vRNA, dims = 1:20)

# DimPlot(vRNA, reduction = "umap")
# DimPlot(vRNA, reduction = "umap", group.by = "CellType")
```

### Score NK cells from de Vries 
For scoring the cells, we used the normalied data from RNA slot.
```{r}
vEXpr <- as.matrix(GetAssayData(vRNA, slot = "data", assay = "RNA"))
```

```{r, message = F, warning = F}
rd_v <- rankGenes(vEXpr)

vRNA$NK_Res <- simpleScore(
  rankData = rd_v,
  upSet =  NK_ResMarker,
  knownDirection = T,
  centerScore = F
)$TotalScore

vRNA$NK_Exh <- simpleScore(
  rankData = rd_v,
  upSet =  NK_ExhMarker,
  knownDirection = T,
  centerScore = F
)$TotalScore

```


### UMAP/PCA using marker genes
```{r}
gg <-  c(NK_ExhMarker, NK_ResMarker)

NKExhTrm_vRNA <- RunPCA(vRNA, features = gg, verbose = F)

NKExhTrm_vRNA <- FindNeighbors(NKExhTrm_vRNA, reduction = "pca", dims = 1:10)
NKExhTrm_vRNA <- FindClusters(NKExhTrm_vRNA, resolution = 0.5)

NKExhTrm_vRNA <- RunUMAP(NKExhTrm_vRNA, reduction = "pca", 
          dims = 1:10,  min.dist = 0.2, n.neighbors = 15)

# DimPlot(NKExhTrm_vRNA, reduction = "umap")
# DimPlot(NKExhTrm_vRNA, reduction = "umap", group.by = "CellType")
# FeaturePlot(NKExhTrm_vRNA, reduction = "umap", features = "NK_Exh")
# FeaturePlot(NKExhTrm_vRNA, reduction = "umap", features = "NK_Res")
```

### UMAP plots coloured by annotations and scores
We colour the UMAPs generated using our Res and Exh marker genes based on different annotations, including the cell annotations (reported by de Vries et al) and Seurat clusters (obtained using our marker genes). Then, we colour the same UMAPs using the Res and Exh scores. Most exhasuted cells are proliferative cells.
```{r, fig.height=4.5, fig.width=4.2, fig.cap="UMAP plots genearted by Res and Exh genes, and coloured by annotations and scores"}
anns <- c("seurat_clusters", "CellType")

# pdf(
#   paste0(figPath, "UMAP_Vries_NK_ExhTrm_RNA_20201005_CellAnnot.pdf"),
#   height = 4.3,
#   width = 4
# )

for(a in anns){
  print(DimPlot(NKExhTrm_vRNA,
                reduction = "umap", 
        group.by = a) +
  theme_minimal() +
  scale_color_manual(values = cols) +
  theme(legend.position = "top") +
  guides(color = guide_legend(
    nrow=2, 
    override.aes = list(size=3))))
}

FeaturePlot(
  NKExhTrm_vRNA,
  "NK_Exh",
  pt.size = 1,
  order = T) +
  scale_color_viridis_c(100) + theme_dark() +
  theme(legend.position = "top")

FeaturePlot(
  NKExhTrm_vRNA,
            "NK_Res",
            pt.size = 1,
            order = T) +
  scale_color_viridis_c(100) + theme_dark() +
  theme(legend.position = "top") 


# dev.off()


```

Blend plot coloured by scores (related to Figure 2 in the paper)
```{r, fig.height=3, fig.width=12}
# pdf(
#   paste0(figPath, "UMAP_PCA_Vries_NK_Blend_ExhRes_RNA_20201005.pdf"),
#   height = 3,
#   width = 12
# )
FeaturePlot(
  NKExhTrm_vRNA,
  features = c("NK_Res", "NK_Exh"),
  pt.size = 1.5,
  order = T,
  blend = T,
  cols =  c("grey90", "orange2", "blue"),
  reduction = "umap"
) 

# dev.off()
```

```{r}
### Save NK data after UMAP/Scores
# saveRDS(vRNA, file = paste0(outPath, "DontPush/deVries_SeuratObj_NK_20201005_UMAP_Scores.RDS"))

```


# James' gut atlas
We downloaded the data rfom Human Gut Atlas by [James et al](https://www.nature.com/articles/s41590-020-0602-z), and generated the Seurat object (see the commnented code below) which includes the data for NK and T cells only. Here we read the Seurat object, and subset to NK cells.

In this data, we need to deal with the variation across patients, as there is a clear donor impact in the data.
```{r}
# counts <- Matrix::readMM(paste0(JamesPath, "Colon_immune_counts.mtx.txt"))
# counts <- t(counts)
# 
# cellAnnot <- read.csv(paste0(JamesPath, "Colon_immune_metadata.csv"), stringsAsFactors = F)
# rownames(cellAnnot) <- cellAnnot$index
# genes <- read.csv(paste0(JamesPath, "Colon_immune_gene_names.csv"), stringsAsFactors = F)
# 
# colnames(counts) <- cellAnnot$index
# rownames(counts) <- genes$index
# 
# 
# gut <-
#   CreateSeuratObject(
#     counts = counts,
#     meta.data = cellAnnot,
#     min.cells = 3,
#     min.features = 200,
#     project = "James_HumanGutAtlas"
#   )
# gut[["percent.mt"]] <- PercentageFeatureSet(gut, pattern = "^MT-")
# 
##---- subset to T and NK cells
# getCells <-
#   c(
#     "Activated CD4 T",
#     "CD8 T",
#     "cycling gd T",
#     "gd T",
#     "ILC",
#     "NK",
#     "Tcm",
#     "Tfh",
#     "Th1",
#     "Th17",
#     "Treg"
#   )
# 
# subGut <- gut[, gut$cell_type %in% getCells]
# 
# # grep("MT-", rownames(gut), value = T)
# subGut[["percent.mt"]] <- PercentageFeatureSet(subGut, pattern = "^MT-")
# saveRDS(subGut, paste0(outPath, "James_HumanGutAtlas_NK_T_SeuratObject.RDS")

# subGut <- readRDS(paste0(JamesPath, "James_HumanGutAtlas_NK_T_SeuratObject.RDS"))
# cd4James <- c("Tfh", "Th1", "Th17", "Treg")
```


We first perfromed QC, normalised the data using SCTransform, and performed all the down stream analysis and we observed clear separation based on donors in the data (as also mentioned by the authors in the James et al paper); therefore, we decided to use Seurat to integrate the data from different donors and perform batch correction. We also tried Harmony (not shown here), but observed a better integration using Seurat.

## Seurat integration
Before doing any normalisation, we run SCTransform on each donor separately.
The main steps are:

"1. Create Seurat object
2. QC by filtering out cells based on percent.mito and nFeature_RNA
3. SCT normalize each dataset specifying the parameter vars.to.regress = percent.mito
4. Integrate all datasets - do not run the ScaleData function after integration
5. Run PCA, UMAP, FindClusters, FindNeighbors (on default assay which is "integrated")
6. Change default assay to "RNA"; normalize then generate FeaturePlots and perform differential expression analysis"

As the integration step takes time, we have commented the code but you can see how we integrated the data. For now, we directly read the RData file containing the integrated data.
```{r}
# subGut_batch <- subGut0[, subGut0$cell_type %in% "NK"]
# subGut_batch <- subGut_batch[, subGut_batch$percent.mt < 20 & subGut_batch$nFeature_RNA < 5000]
# 
# donor.list <- SplitObject(subGut_batch, split.by = "donor")
# 
# 
# donor.list <- lapply(X = donor.list, FUN = function(x) {
#  SCTransform(
#   object = x,
#   assay = "RNA",
#   new.assay.name = "SCT",
#   do.correct.umi = TRUE,
#   ncells = NULL,
#   variable.features.n = 5000,
#   vars.to.regress = "percent.mt",  ## NULL by default
#   do.scale = FALSE,
#   do.center = TRUE,
#   conserve.memory = FALSE,
#   return.only.var.genes = FALSE,
#   seed.use = 20201005,
#   verbose = TRUE
# )
# })
# 
# # options(future.globals.maxSize = 4000000000)
# 
# donor.features <-
#   SelectIntegrationFeatures(object.list = donor.list, nfeatures = 5000)
# 
# donor.list <-
#   PrepSCTIntegration(object.list = donor.list,
#                      anchor.features = donor.features,
#                      verbose = TRUE)
# 
# donor.anchors <-
#   FindIntegrationAnchors(
#     object.list = donor.list,
#     normalization.method = "SCT",
#     anchor.features = donor.features,
#     verbose = TRUE, 
#     k.anchor = 5, k.filter = 10, k.score = 10,
#     dims= 1:10 ## has to be smaller than cell number in batches
#   )
# 
# subGut_integrated <-
#   IntegrateData(
#     anchorset = donor.anchors,
#     normalization.method = "SCT",
#     verbose = TRUE, 
#     k.weight = 10,
#     dims = 1:10
#   )
# 
# 
# saveRDS(subGut_integrated, file = paste0(JamesPath, "James_HumanGutAtlas_NK_IntegratedSCT.RDS"))

subGut_integrated <- readRDS(paste0(JamesPath, "James_HumanGutAtlas_NK_IntegratedSCT.RDS"))
```

#### PCA, UMAP, and Clustering
We use the corrected data for PCA and UMAP
```{r}
DefaultAssay(subGut_integrated) <- "integrated"

subGut_integrated <- RunPCA(subGut_integrated, npcs = 20, verbose = F)

subGut_integrated <-
  RunUMAP(
    subGut_integrated,
    reduction = "pca",
    dims = 1:20,
    n.neighbors = 20,
    min.dist = 0.2,
    seed.use = 20201005
  )

subGut_integrated <-
  FindNeighbors(subGut_integrated, reduction = "pca", dims = 1:20)

subGut_integrated <-
  FindClusters(subGut_integrated, resolution = 0.5)

```



### Score cells
Score cells using RNA assay
```{r}
DefaultAssay(subGut_integrated) <- "RNA"
# Normalize RNA data for visualization purposes
subGut_integrated <- NormalizeData(subGut_integrated, verbose = FALSE)
```
 
```{r, message = F, warning = F}

exprDataJ_RNA <-
  as.matrix(GetAssayData(subGut_integrated, slot = "data", assay = "RNA"))

rd_JRNA <- rankGenes(exprDataJ_RNA)

subGut_integrated$NK_Res <- simpleScore(
  rankData = rd_JRNA,
  upSet = NK_ResMarker,
  knownDirection = T,
  centerScore = F
)$TotalScore

subGut_integrated$NK_Exh <- simpleScore(
  rankData = rd_JRNA,
  upSet = NK_ExhMarker,
  knownDirection = T,
  centerScore = F
)$TotalScore

```


### UMAP/PCA using marker genes
```{r}

gg <- c(NK_ExhMarker, NK_ResMarker)

subGut_integrated2 <- subGut_integrated
DefaultAssay(subGut_integrated2) <- "integrated"

subGut_integrated2 <-
  RunPCA(
    subGut_integrated2,
    verbose = F,
    features = gg
  )

subGut_integrated2 <-
  RunUMAP(
    subGut_integrated2,
    reduction = "pca",
    dims = 1:20,
    n.neighbors = 20,
    min.dist = 0.2,
    seed.use = 20201005
  )

subGut_integrated2 <- FindNeighbors(subGut_integrated2, reduction = "pca", dims = 1:20)
subGut_integrated2 <- FindClusters(subGut_integrated2, resolution = 0.4)

```

We colour the UMAPs generated using our Res and Exh marker genes based on different annotations, including the cell annotations (reported by de Vries et al) and Seurat clusters (obtained using our marker genes). Then, we colour the same UMAPs using the Res and Exh scores. Most exhausted cells are proliferative cells.
```{r, fig.height=4.5, fig.width=4.2, fig.cap="UMAP plots genearted by Res and Exh genes, and coloured by annotations and scores"}
anns <- c("seurat_clusters", "cell_type", "region", "gender", "donor")

# pdf(
#   paste0(figPath, "UMAP_James_NK_ExhTrmMarkers_SeuratIntegratedSCT_CellAnnot.pdf"),
#   height = 4.3,
#   width = 4
# )

for(a in anns){
  print(DimPlot(subGut_integrated2, 
                reduction = "umap",
        group.by = a) +
  theme_minimal() +
  scale_color_manual(values = cols) +
  theme(legend.position = "top") +
  guides(color = guide_legend(
    nrow=2, 
    override.aes = list(size=3))))
}

FeaturePlot(subGut_integrated2,
            "NK_Exh",
            reduction = "umap",
            pt.size = 1,
            order = T) +
  scale_color_viridis_c(100) + theme_dark() +
  theme(legend.position = "top")

FeaturePlot(subGut_integrated2,
            "NK_Res",
            reduction = "umap",
            pt.size = 1,
            order = T) +
  scale_color_viridis_c(100) + theme_dark() +
  theme(legend.position = "top")

# dev.off()

```

Blend plot coloured by scores (related to the Figure 2 in the paper):
```{r, fig.height=3, fig.width=12}
# pdf(
#   paste0(figPath, "UMAP_James_Blend_NK_ExhResMarkers_SeuratIntegratedSCT.pdf"),
#   height = 3,
#   width = 12
# )
FeaturePlot(subGut_integrated2,
            features = c("NK_Res", "NK_Exh"),
            pt.size = 1,
            order = T, blend = T,
            cols =  c("grey90", "orange2", "blue")) +
  scale_color_viridis_c(100) + theme_dark() +
  theme(legend.position = "top") 
# dev.off()
```

# Session information
Here are the list of tools and packages (along with their versions) used in this document. 
```{r}
sessionInfo()
```

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 598px;"></div>